---
title: "RepeatCounting_contigs"
output: html_document
---

```{r}
# Required packages 
library(ggplot2)
library(gridExtra)
library(ggthemes)
library(Hmisc)
library(dplyr)
library(ggpubr)

# Define initial parameters
workdir <- '/Volumes/abt6_projects8/1001g_plus_scaffolding/code/GitHub_1001gPLUS_FERNANDO/'
WINDOW <- 1000
CUTOFF_CONTIG <- 10000

setwd(workdir)

# Load accession names from universal list
files <- read.table(paste(workdir, '27accessions.txt', sep = ''))
accs <- files$V1
accs

# Remove TAIR10 from the pack
accs <- accs[-grep("TAIR10", accs)]
accs

```

```{r}
# Define chromosomes
chromosomes <- data.frame(Chr=c('Chr1', 'Chr2', 'Chr3', 'Chr4', 'Chr5'), Size=c(30427671, 19698289, 23459830, 18585056, 26975502), stringsAsFactors=FALSE)

# Define repeat lengths
repeat_length <- data.frame(Type=c('centromere', '5S', '45S', 'telomere', 'mitochondria', 'chloroplast', "repeat_region", "ClassI_LTR" ,"ClassII_Helitrons" ,"ClassI_nonLTR" ,"ClassII_TIR" ), Size=c(178, 120, 9000, 7, 1, 1, 1, 1, 1, 1, 1), stringsAsFactors=FALSE) 

# Initialize a counter data.frame that will collect Frequencies of repeats at Contig edges from all accessions
ALL_edges <- data.frame(ALL_edges_TMP=NA, Freq=NA, Accession=NA)

PLACED_edges <- data.frame(PLACED_edges_TMP=NA, Freq=NA, Accession=NA)
UNPLACED_edges <- data.frame(UNPLACED_edges_TMP=NA, Freq=NA, Accession=NA)

# Initialize a counter for genome size, and assembly and unassembly lengths  
genome_sizes <- data.frame(Accession=NA, Genome_size=NA, Assembled_length=NA , Unassembled_length=NA )

# Initialize a counter data.frame that will collect Repeat length from all accessions   
COUNT_length_PLACED <- data.frame(Type=NA, Length=NA, Accession=NA)
COUNT_length_UNPLACED <- data.frame(Type=NA, Length=NA, Accession=NA)

# Loop for accessions
for(i in 1:length(accs)){

  ##################################################################################
  ############################## READ & FORMAT inputs ##############################

  # Read Genome Size
  genome <- read.table(paste(workdir, 'output_findGSE/', accs[i], '/R_output/v1.94.est.', accs[i], '.21mer.histo.genome.size.estimated.k21to21.fitted.txt', sep = ''),skip=4,nrow=1)
  genome[1,1] <- accs[i]
  colnames(genome) <- c('Accession', 'Genome_size')
  
  # Read faidx with ALL contig lengths
  faidx <- read.table(paste(workdir, '3_RepeatAnnotation_contigs_v2.1/', accs[i], '/', accs[i], '.fa.fai', sep = ''))
  faidx <- faidx[,-c(3:5)]
  faidx$V1 <- gsub("\\|arrow\\|pilon", "" , faidx$V1)
  colnames(faidx) <- c('Contig', 'Contig_length')

  ######## DEPRECATED when new pre-defined contigs became available ################
  ## Read list of Chr and UNPLACED contigs, and keep only UNPLACED
  #faidx_UNPLACED <- read.table(paste(workdir, '2_Ragtag_masked_150kb_q60_f10000_i0.6/', accs[i], '/', accs[i], '.ragtag_scaffolds.fa.fai', sep = ''))
  #faidx_UNPLACED <- faidx_UNPLACED[-grep("^Chr", faidx_UNPLACED$V1), -c(3:5)]
  #faidx_UNPLACED$V1 <- gsub("\\|arrow\\|pilon", "" , faidx_UNPLACED$V1)
  #colnames(faidx_UNPLACED) <- c('Contig', 'Contig_length')
  #
  ## Define placed contigs
  #faidx_PLACED <- faidx[ !(faidx$Contig %in% faidx_UNPLACED$Contig), ]
  ######## DEPRECATED ##############################################################   

  
  # Define placed contigs
  faidx_PLACED <- faidx[grep("_p", faidx$Contig), ]
  
  # Define un-placed contigs
  faidx_UNPLACED <- faidx[grep("_t", faidx$Contig), ]
  
  # Read and transform GFF3 file (this should change when I have TEs)
  gff <- read.table(paste(workdir, '3_RepeatAnnotation_contigs_v2.1/', accs[i], '/', accs[i], '.Repeats_TEanno.gff3', sep = ''))
  
  head(gff)
  
  gff$Type <- unlist(lapply(strsplit(as.character(gff$V9), "[=_]"), "[", 2))
  # Remove overlapping TE features
  gff <- subset(gff, Type != "repeat" & Type != "lTSD" & Type != "lLTR" & Type != "rLTR" & Type != "rTSD")
  # Dissect TEs classifications 
  gff$Type <- mapply(FUN = function(V3, Type){
    if(Type == "TE" | Type == "LTRRT" ){Type <- V3}
    else{Type <- Type}}, gff$V3, gff$Type )
  
  gff$Length <- unlist(lapply(strsplit(as.character(gff$V9), "Length="), "[", 2))
  gff$Length <- as.numeric(unlist(lapply(strsplit(as.character(gff$Length), "[;]"), "[", 1)))
  head(gff)
  
  gff$Divergence <- unlist(lapply(strsplit(as.character(gff$V9), "Divergence="), "[", 2))
  gff$Divergence <- as.numeric(unlist(lapply(strsplit(as.character(gff$Divergence), "[;]"), "[", 1)))
  head(gff)
  
  gff$Identity <- unlist(lapply(strsplit(as.character(gff$V9), "Identity="), "[", 2))
  gff$Identity <- as.numeric(unlist(lapply(strsplit(as.character(gff$Identity), "[;]"), "[", 1)))
  head(gff)

  gff$V8 <- unlist(lapply(strsplit(as.character(gff$V9), "Coordinates="), "[", 2))
  gff$Organelle_START <- as.numeric(unlist(lapply(strsplit(as.character(gff$V8), "[-;]"), "[", 1)))
  gff$Organelle_END <- as.numeric(unlist(lapply(strsplit(as.character(gff$V8), "[-;]"), "[", 2)))
  head(gff)
  
  # Get rid of unnecessary columns
  gff <- gff[ , -c(2,3,6,7,8,9)]
  colnames(gff)[1:3] <- c('Contig', 'Start', 'End')
  head(gff)

  # CORRECT Length for the actual one in assembly (this way also TEs play)
  gff$Length <- gff$End - gff$Start

  # Mask with NA's repeats shorter than 90% their expected size. (too harsh for some large repeats)
  gff <- merge(gff, repeat_length, all.x = TRUE)
  gff$Start.2 <- mapply(FUN = function(Start, Length, Size){if(is.na(Length/Size)){Start <- NA}else if(Length/Size >= 0.9){Start <- Start}else{Start <- NA}}, gff$Start, gff$Length,  gff$Size )
  gff$End.2 <- mapply(FUN = function(End, Length, Size){if(is.na(Length/Size)){End <- NA}else if(Length/Size >= 0.9){End <- End}else{End <- NA}}, gff$End, gff$Length,  gff$Size )
  
  # Create unique identifier for each event
  gff$ID_start <- paste(unlist(lapply(strsplit(as.character(gff$Contig), "[_|]"), "[", 2)), gff$Start, sep="-")
  gff$ID_end <- paste(unlist(lapply(strsplit(as.character(gff$Contig), "[_|]"), "[", 2)), gff$End, sep="-")

  # EXTRA # Filter out duplicted Chr hits due to internal repeated structure of chloroplasts
  #gff <- gff[ -which(duplicated(gff$ID_start)), ]
  
  ##################################################################################
  ############################## Assembly LENGTH ##############################
  
  genome$Assembled_length <- sum( faidx_PLACED$Contig_length )
  genome$Unassembled_length <- sum( faidx_UNPLACED$Contig_length )
  genome_sizes <- rbind(genome_sizes, genome)
  
  ##################################################################################
  ############################## Repeats LENGTH ##############################

  #Assembled 
  gff_placed <- gff[ gff$Contig %in% faidx_PLACED$Contig, ]
  gff_placed <- aggregate(Length ~ Type, gff_placed, sum)
  gff_placed$Accession <- accs[i]
  # Add length of uncharacterized fraction
  gff_placed[dim(gff_placed)[1]+1 , 2] <- genome[3] - sum(gff_placed$Length)
  gff_placed[dim(gff_placed)[1] , 1] <- 'Nonrepetitive_genome'
  gff_placed[dim(gff_placed)[1] , 3] <- accs[i]
  COUNT_length_PLACED <- rbind(COUNT_length_PLACED, gff_placed)
  
  #Unassembled
  gff_unplaced <- gff[ gff$Contig %in% faidx_UNPLACED$Contig, ]
  gff_unplaced <- aggregate(Length ~ Type, gff_unplaced, sum)
  gff_unplaced$Accession <- accs[i]
  # Add length of uncharacterized fraction
  gff_unplaced[dim(gff_unplaced)[1]+1 , 2] <- genome[4] - sum(gff_unplaced$Length)
  gff_unplaced[dim(gff_unplaced)[1] , 1] <- 'Nonrepetitive_genome'
  gff_unplaced[dim(gff_unplaced)[1] , 3] <- accs[i]
  COUNT_length_UNPLACED <- rbind(COUNT_length_UNPLACED, gff_unplaced)
  
  ##################################################################################
  ############################## Contig BREAKS ##############################

  # Find FIRST event per Contig
  TEMP_s <- aggregate(gff$Start, by=list(Type=gff$Contig), FUN=min)
  colnames(TEMP_s) <- c("Contig", "Chr_start")
  TEMP_s$ID_start <- paste(unlist(lapply(strsplit(as.character(TEMP_s$Contig), "[_|]"), "[", 2)), TEMP_s$Chr_start, sep="-")
  TEMP_s <- TEMP_s[ , -1 ]
  TEMP_s <- merge(TEMP_s, gff, by="ID_start")
  TEMP_s <- TEMP_s[ , -c(10:15) ]
  
  # Find LAST event per Contig  
  TEMP_e <- aggregate(gff$End, by=list(Type=gff$Contig), FUN=max)
  colnames(TEMP_e) <- c("Contig", "Chr_end")
  TEMP_e$ID_end <- paste(unlist(lapply(strsplit(as.character(TEMP_e$Contig), "[_|]"), "[", 2)), TEMP_e$Chr_end, sep="-")
  TEMP_e <- TEMP_e[ , -1 ]
  TEMP_e <- merge(TEMP_e, gff, by="ID_end")
  TEMP_e <- TEMP_e[ , -c(10:15) ]
  
  # Combine both coordinates of each contig into a single data frame  
  TEMP_se <- merge(TEMP_s, TEMP_e, by="Contig")
  
  # Integrate Total Contig length (Consider to dump very short Contigs)
  TEMP_se <- merge(TEMP_se, faidx, by="Contig", all.y=TRUE)
  head(TEMP_se)
  
  # Mark repeat occurrences at contig edges
  TEMP_se$Chr_start <- mapply(function(A, B){ 
    if(is.na(A)){A <- NA}
    else if(A < WINDOW){A <- B}
    else{A<- NA}}, TEMP_se$Chr_start, TEMP_se$Type.x)

  TEMP_se$Chr_end <- mapply(function(C, D, E){ 
    if(is.na(C)){C <- NA}
    else if(C > E - WINDOW){C <- D}
    else{C <- NA}}, TEMP_se$Chr_end, TEMP_se$Type.y, TEMP_se$Contig_length)
  
  # Throw out CONTIGS below a certain size for further analysis
  TEMP_se <- subset(TEMP_se, Contig_length > CUTOFF_CONTIG)
  
  # Count and organize those occurrences in a Frequency table
  ALL_edges_TMP <- c(TEMP_se$Chr_start, TEMP_se$Chr_end)
  ALL_edges_TMP <- data.frame(table(ALL_edges_TMP, exclude = NULL))
  ALL_edges_TMP$Accession <- accs[i]

  # Divide ALL contigs in PLACED and UNPLACED
  placed_TMP <- TEMP_se[ !(TEMP_se$Contig %in% faidx_UNPLACED$Contig), ]
  unplaced_TMP <- TEMP_se[ TEMP_se$Contig %in% faidx_UNPLACED$Contig, ]
  
  # Count and organize those occurrences in a Frequency table
  PLACED_edges_TMP <- c(placed_TMP$Chr_start, placed_TMP$Chr_end)
  PLACED_edges_TMP <- data.frame(table(PLACED_edges_TMP, exclude = NULL))
  PLACED_edges_TMP$Accession <- accs[i]
  
  UNPLACED_edges_TMP <- c(unplaced_TMP$Chr_start, unplaced_TMP$Chr_end)
  UNPLACED_edges_TMP <- data.frame(table(UNPLACED_edges_TMP, exclude = NULL))
  UNPLACED_edges_TMP$Accession <- accs[i]

  # Concatenate stuff
  ALL_edges <- rbind(ALL_edges, ALL_edges_TMP)
  PLACED_edges <- rbind(PLACED_edges, PLACED_edges_TMP)
  UNPLACED_edges <- rbind(UNPLACED_edges, UNPLACED_edges_TMP)

  
}

genome_sizes <- genome_sizes[-1, ]

COUNT_length_PLACED <- COUNT_length_PLACED[-1, ]
COUNT_length_UNPLACED <- COUNT_length_UNPLACED[-1, ]

ALL_edges <- ALL_edges[-1, ]
PLACED_edges <- PLACED_edges[-1, ]
UNPLACED_edges <- UNPLACED_edges[-1, ]



  ########################  ########################  ########################  ########################  ########################  ########################
 
ALL_edges_ <- ALL_edges 
PLACED_edges_ <- PLACED_edges
UNPLACED_edges_ <- UNPLACED_edges

COUNT_length_PLACED_ <- COUNT_length_PLACED
COUNT_length_UNPLACED_ <- COUNT_length_UNPLACED


genome_sizes
```

# BLAST-based repeats

```{r}
# Create OUTPUT directory for ALL plots
dir.create(paste(workdir, '4_plots', sep = ''), showWarnings = FALSE)

# Create OUTPUT directory for ALL repeat files
dir.create(paste(workdir, '5_output_repeats', sep = ''), showWarnings = FALSE)

#Create empty data frame
blast_results <- data.frame(Accession=NA, Centromeres=NA, rDNA_45S=NA, rDNA_5S=NA, Kmer_coverage=NA, Genome_size.again=NA)

# Read CEN, 45S and 5S repeats
for(i in 1:length(accs)){
  blast_results[i,] <- read.table(paste0(workdir, 'output_blast/results/', accs[i], '.f_ALL.blastn.REPEATS.txt'))
}

blast_results <- blast_results[,-6]

# Repeats total
blast_results$Repeats_total <- blast_results$Centromeres + blast_results$rDNA_45S + blast_results$rDNA_5S

# Substitute letter names for ecotype IDs before merging
#blast_results$Accession <- gsub("85-3", "22001", blast_results$Accession)
#blast_results$Accession <- gsub("35-1", "22002", blast_results$Accession)
#blast_results$Accession <- gsub("Taz-0", "22003", blast_results$Accession)
#blast_results$Accession <- gsub("Elh-2", "22004", blast_results$Accession)
#blast_results$Accession <- gsub("R1", "22005", blast_results$Accession)
#blast_results$Accession <- gsub("A1", "22006", blast_results$Accession)
#blast_results$Accession <- gsub("ET86-4", "22007", blast_results$Accession)


# MErge with GEnome Size estimates
blast_results$Accession <- sapply(strsplit(blast_results$Accession, "[.]"), "[", 1)
blast_results <- merge(genome_sizes, blast_results, by = "Accession")


blast_results

# Export summary GENOME SIZE and REPEATS
write.table(blast_results, paste0(workdir, '/5_output_repeats/SuppTable.genome_size_and_repeats.27accs.txt'), row.names = FALSE, quote = FALSE, sep = "\t")

# Generate PLOT
plot_cor_repeats <- ggplot(blast_results, aes(x=(Genome_size - Assembled_length)/1000000, y=Repeats_total/1000000)) +
  geom_point() + 
  geom_text(label=blast_results$Accession) + 
  theme_minimal() +
  xlab("Genome size - Chromosome scaffolds (Mb)") + ylab("Centromeres + 45S rDNAs + 5S rDNAs  (Mb)") +
  stat_summary(fun.data=mean_cl_normal) + 
  geom_smooth(method='lm', formula= y~x) +
  stat_cor(aes(label=..rr.label..), label.x=32.5, label.y=25.5, size=4.2, digits=2)

print(plot_cor_repeats)

# PDF
pdf(paste0(workdir, '/4_plots/SuppFig.Unreachable_vs_Repeats.27accs.pdf'), width = 8, height = 6)
plot_cor_repeats
dev.off()

# PNG
png(paste0(workdir, '/4_plots/SuppFig.Unreachable_vs_Repeats.27accs.png'), width = 160, height = 120, units='mm', res = 300)
plot_cor_repeats
dev.off()

# Correlation stats
summary(lm(blast_results$Repeats_total ~ (blast_results$Genome_size - blast_results$Assembled_length)))
# Adjusted R-squared:  0.9418 
# p-value: < 2.2e-16


print(plot_cor_repeats)

```


# PLOTS


# COLOR palette

None                              '#969696'
45S                               '#ce1256'
5S                                '#c994c7'  
centromere                        '#67001f'
Transposable element              '#3690c0'
Other repeat                      '#bdbdbd'   '#4d4d4d'
Telomere                          '#8856a7'
Nuclear organellar insertion      "#bf812d"
Nuclear mitochondria insertion    "#8c510a"
Nuclear plastid insertion         '#dfc27d'

Estimated non-scaffolded genome   '#980043'
Non-repetitive genome             '#1a9846'





Summary of ALL accessions in one plot. 
```{r}
PLACED_edges_ <- PLACED_edges
# Edit each data frame
PLACED_edges_$Assembly <- 'Chromosome scaffolds'
colnames(PLACED_edges_)[1] <- 'Annotation'
  
UNPLACED_edges_$Assembly <- 'Unplaced contigs'
colnames(UNPLACED_edges_)[1] <- 'Annotation'
  
# Concatenate both contig sets
edges_ <-  rbind(PLACED_edges_, UNPLACED_edges_)
dim(edges_)

# Keep TEs by class, just in case
edges_$Annotation2 <- edges_$Annotation

edges_

# Substitute annotations names for plot
edges_$Annotation <- gsub("centromere", "Centromere", edges_$Annotation)
edges_$Annotation <- gsub("45S", "45S rDNA", edges_$Annotation)
edges_$Annotation <- gsub("^5S", "5S rDNA", edges_$Annotation)
edges_$Annotation <- gsub("ClassI_nonLTR", "Transposable element", edges_$Annotation)
edges_$Annotation <- gsub("ClassI_LTR", "Transposable element", edges_$Annotation)
edges_$Annotation <- gsub("ClassII_TIR", "Transposable element", edges_$Annotation)
edges_$Annotation <- gsub("ClassII_Helitrons", "Transposable element", edges_$Annotation)
edges_$Annotation <- gsub("mitochondria", "Mitochondria", edges_$Annotation)
edges_$Annotation <- gsub("chloroplast", "Plastid", edges_$Annotation)
edges_$Annotation <- gsub("repeat_region", "Other repeat", edges_$Annotation)
edges_$Annotation <- gsub("telomere", "Telomere", edges_$Annotation)
edges_$Annotation[ which(is.na(edges_$Annotation)) ] <- "None"

# Export summary CONTIG BREAKS
write.table(edges_, paste0(workdir, '/5_output_repeats/SuppTable.Contig_breaks.txt'), row.names = FALSE, quote = FALSE, sep = "\t")

# Define colours
ELEMENTS <- c( "Centromere", "5S rDNA", "45S rDNA", "Other repeat","Transposable element",  "Mitochondria", "Plastid", "Telomere", "None")
myColors_ALL <-  c('#67001f', '#c994c7', '#ce1256', '#bdbdbd', '#3690c0', "#8c510a", "#dfc27d", '#8856a7', '#1c1c1c')
names(myColors_ALL) <- as.factor(ELEMENTS)


# Re-order repeats within each column
aggregate(edges_$Freq, by = list(edges_$Annotation, edges_$Assembly), sum)
edges_$Annotation <- factor(edges_$Annotation, levels=c("None",  "Plastid", "Mitochondria", "Other repeat", "Transposable element", "45S rDNA", "5S rDNA", "Centromere", "Telomere"), exclude=NULL)   

# Generate PLOT
plot_contig_breaks <- ggplot(edges_, aes(fill=Annotation, y=Freq, x=Assembly)) + 
  geom_bar(position="fill", stat="identity") +
  scale_fill_manual( values = myColors_ALL,  na.value="#969696" ) +
  ylab("Frequency") +
  theme(axis.text.x = element_text(angle = 0, vjust = -1, hjust=0.5)) +
  theme_minimal() +
  theme(legend.position = "right", text = element_text(size=18), axis.text.y = element_text(angle = 0, vjust = 0.5, hjust=1, size=15), axis.text.x = element_text(size=15))

#  ggtitle(label = "Annotation of Contig breaks\nSCAFFOLDED vs UNPLACED") 

print(plot_contig_breaks)
#  facet_grid(~Accession)

# PDF
#pdf(paste0('/Users/frabanal/Downloads/SuppFig.ContigBreaks_summary.pdf'), width = 10, height = 8)
pdf(paste0(workdir, '/4_plots/SuppFig.ContigBreaks_summary.pdf'), width = 10, height = 8)
plot_contig_breaks
dev.off()

# PNG
#png(paste0('/Users/frabanal/Downloads/SuppFig.ContigBreaks_summary.png'), width = 250, height = 200, units='mm', res = 300)
png(paste0(workdir, '/4_plots/SuppFig.ContigBreaks_summary.png'), width = 250, height = 200, units='mm', res = 300)
plot_contig_breaks
dev.off()

print(plot_contig_breaks)
```

Plot all repeat lengths in the context of genome size
```{r}
# K-mer estimate of Genome size MINUS the lenfth of the scaffolded chromosomes 
append_genome <- data.frame( 'Unreachable_genome', genome_sizes$Genome_size - genome_sizes$Assembled_length, genome_sizes$Accession)
colnames(append_genome) <- c("Type", "Length", "Accession")

# Concatenate genome size to assembled part
COUNT_length_PLACED_ <- rbind(COUNT_length_PLACED_, append_genome)

# Order data set by genome size level
mayor2 <- genome_sizes[,1:2]
mayor2 <- mayor2[sort(mayor2[,2], index.return=TRUE)$ix, ]
COUNT_length_PLACED_$Accession <- factor(COUNT_length_PLACED_$Accession , levels=factor(mayor2[,1]) )   

# Keep TEs by class, just in case
COUNT_length_PLACED_$Type2 <- COUNT_length_PLACED_$Type

# Substitute name
#COUNT_length_PLACED_$Type <- gsub("centromere", "Centromere", COUNT_length_PLACED_$Type)
COUNT_length_PLACED_$Type <- gsub("centromere", "Centromere", COUNT_length_PLACED_$Type)
#COUNT_length_PLACED_$Type <- gsub("45S", "45S rDNA", COUNT_length_PLACED_$Type)
COUNT_length_PLACED_$Type <- gsub("45S", "45S rDNA", COUNT_length_PLACED_$Type)
#COUNT_length_PLACED_$Type <- gsub("^5S", "5S rDNA", COUNT_length_PLACED_$Type)
COUNT_length_PLACED_$Type <- gsub("^5S", "5S rDNA", COUNT_length_PLACED_$Type)
COUNT_length_PLACED_$Type <- gsub("ClassI_nonLTR", "Transposable element", COUNT_length_PLACED_$Type)
COUNT_length_PLACED_$Type <- gsub("ClassI_LTR", "Transposable element", COUNT_length_PLACED_$Type)
COUNT_length_PLACED_$Type <- gsub("ClassII_TIR", "Transposable element", COUNT_length_PLACED_$Type)
COUNT_length_PLACED_$Type <- gsub("ClassII_Helitrons", "Transposable element", COUNT_length_PLACED_$Type)
COUNT_length_PLACED_$Type <- gsub("mitochondria", "Nuclear organellar insertion", COUNT_length_PLACED_$Type)
COUNT_length_PLACED_$Type <- gsub("chloroplast", "Nuclear organellar insertion", COUNT_length_PLACED_$Type)
COUNT_length_PLACED_$Type <- gsub("repeat_region", "Other repeat", COUNT_length_PLACED_$Type)
COUNT_length_PLACED_$Type <- gsub("telomere", "Telomere", COUNT_length_PLACED_$Type)
COUNT_length_PLACED_$Type <- gsub("Nonrepetitive_genome", "Non-repetitive genome", COUNT_length_PLACED_$Type)
COUNT_length_PLACED_$Type <- gsub("Unreachable_genome", "Estimated non-scaffolded genome", COUNT_length_PLACED_$Type)

# Export summary GENOME SIZE and REPEATS
write.table(COUNT_length_PLACED_, paste0(workdir, '/5_output_repeats/SuppTable.Repeats_scaffolded_genome.txt'), row.names = FALSE, quote = FALSE, sep = "\t")

# Define colours
#ELEMENTS_ <- c("Estimated non-scaffolded genome", "Non-repetitive genome", "Nuclear mitochondria insertion", "Nuclear plastid insertion", "Telomere", "45S rDNA", "5S rDNA", "Centromere", "Other repeat",  "Transposable element")
#myColors_TEs_ <-  c('#969696', '#d4b9da',   "#bf812d", "#8c510a", '#8856a7', '#67001f', '#b30000', '#252525', '#c7eae5', '#01998d')
ELEMENTS_ <- c("Estimated non-scaffolded genome", "Non-repetitive genome", "Nuclear organellar insertion", "Telomere", "45S rDNA", "5S rDNA", "Centromere", "Other repeat",  "Transposable element")
myColors_ALL_ <-  c('#393939', '#1a9846',   "#bf812d", '#8856a7', '#ce1256','#c994c7',  '#67001f', '#bdbdbd', '#3690c0')
names(myColors_ALL_) <- as.factor(ELEMENTS_)

#BLUE # "Average" color: #3690c0

#RED  # "Average" color for non-scaffolded: #980043 
# 45S 5S CEN: '#ce1256','#c994c7',  '#67001f'


# Re-order repeats within each column
COUNT_length_PLACED_$Type <- factor(COUNT_length_PLACED_$Type, levels=c(ELEMENTS_), exclude=NULL)   
      

# Generate PLOT 
# Plot Proportion of Repeats in the Assembled scaffolds 
plot_repeats_scaffolds <- ggplot(COUNT_length_PLACED_, aes(fill=Type, y=Length/1000000, x=Accession)) + 
  geom_bar(position="stack", stat="identity") +
  scale_fill_manual("Annotation", values = myColors_ALL_,  na.value="#d4b9da" ) +
  theme_minimal() +
  theme(text = element_text(size=20), axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1, size=15), axis.text.y = element_text(size=15)) +
  ggtitle(label = "Total contribution of Repeats \nScaffolded fraction of the assembly") +
  xlab("Accession") + ylab("Cumulative length (Mb)") 


plot_repeats_scaffolds


```


Plot with zero-baseline for better illustration purposes

```{r}
# Covert to negative 
COUNT_length_PLACED_NEW <- COUNT_length_PLACED_ %>% 
  # a trick!
  mutate(Length = if_else(Type == "Centromere" | Type == "45S rDNA" | Type == "5S rDNA" | Type == "Nuclear organellar insertion" | Type == "Transposable element" | Type == "Other repeat" | Type == "Telomere" | Type == "Non-repetitive genome" | Type == "NA", -Length, Length)) 

# Prepare object for "geom_text" aesthics
# Recover Non-scafolded genome and merge it with genome_sizes
NonScaffolded <- subset(COUNT_length_PLACED_, Type == "Estimated non-scaffolded genome")

genome_sizes_NEW <- merge(genome_sizes, NonScaffolded, by = "Accession")

# Generate PLOT
plot_repeats_zero <- 
  COUNT_length_PLACED_NEW %>% 
  ggplot(aes(y = Accession, x = Length/1000000, group = Type, fill = Type)) +
  geom_bar(position="stack", stat = "identity", width = 0.85) +
  #scale_y_discrete(limits = the_order) +
  # another trick!
  scale_x_continuous(limits = c(-122, 42), breaks = seq(-120, 40, 20), labels = abs(seq(-120, 40, 20))) +
  labs(x = "Cumulative length (Mb)", y = "Accession", title = "")  +
  scale_fill_manual("Annotation", values = myColors_ALL_,  na.value="#d4b9da") +
  theme_minimal() +
  theme(legend.position = "none", text = element_text(size=18), axis.text.y = element_text(angle = 0, vjust = 0.5, hjust=1, size=15), axis.text.x = element_text(size=15)) +
  geom_text(data = genome_sizes_NEW, aes(y= Accession, x = Length/1000000, label = paste0(format(round(Genome_size/1000000, digits = 1), nsmall=1), " ")), inherit.aes = FALSE, angle = 0, vjust = 0.5, hjust = -0.2, size=5) 

print(plot_repeats_zero)


# PDF
#pdf(paste0('/Users/frabanal/Downloads/MainFig.Genome_size_at_zero.pdf'), width = 10, height = 8)
pdf(paste0(workdir, '/4_plots/MainFig.Genome_size_at_zero.pdf'), width = 10, height = 8)
plot_repeats_zero
dev.off()

# PNG
#png(paste0('/Users/frabanal/Downloads/MainFig.Genome_size_at_zero.png'), width = 250, height = 200, units='mm', res = 300)
png(paste0(workdir, '/4_plots/MainFig.Genome_size_at_zero.png'), width = 250, height = 200, units='mm', res = 300)
plot_repeats_zero
dev.off()

print(plot_repeats_zero) 


```


```{r}

plot_legends <- plot_repeats_zero +
  theme(legend.position = "right", text = element_text(size=18), axis.text.y = element_text(angle = 0, vjust = 0.5, hjust=1, size=15), axis.text.x = element_text(size=15))

# PDF
#pdf(paste0('/Users/frabanal/Downloads/MainFig.Genome_size_at_zero.pdf'), width = 10, height = 8)
pdf(paste0(workdir, '/4_plots/MainFig.Genome_size_at_zero.Legends.pdf'), width = 10, height = 8)
plot_legends
dev.off()

# PNG
#png(paste0('/Users/frabanal/Downloads/MainFig.Genome_size_at_zero.png'), width = 250, height = 200, units='mm', res = 300)
png(paste0(workdir, '/4_plots/MainFig.Genome_size_at_zero.Legends.png'), width = 250, height = 200, units='mm', res = 300)
plot_legends
dev.off()

print(plot_legends) 



```

Plot contribution of repeats SCAFFOLDED vs UNPLACED contigs

```{r}
# Add column to define whether scaffolded or unplaced
COUNT_length_PLACED$Source <- "Chromosome scaffolds"
COUNT_length_UNPLACED$Source <- "Unplaced contigs"

# Concatenate genome size to assembled part
COUNT_length <- rbind(COUNT_length_PLACED, COUNT_length_UNPLACED)

# Order data set by genome size level defined before
COUNT_length$Accession <- factor(COUNT_length$Accession , levels=factor(mayor2[,1]) )   

# Keep TEs by class, just in case
COUNT_length$Type2 <- COUNT_length$Type

# Substitute name
COUNT_length$Type <- gsub("centromere", "Centromere", COUNT_length$Type)
COUNT_length$Type <- gsub("45S", "45S rDNA", COUNT_length$Type)
COUNT_length$Type <- gsub("^5S", "5S rDNA", COUNT_length$Type)
COUNT_length$Type <- gsub("polinton", "Transposable element", COUNT_length$Type)
COUNT_length$Type <- gsub("non_ClassI_LTR", "Transposable element", COUNT_length$Type)
COUNT_length$Type <- gsub("ClassI_nonLTR", "Transposable element", COUNT_length$Type)
COUNT_length$Type <- gsub("ClassI_LTR", "Transposable element", COUNT_length$Type)
COUNT_length$Type <- gsub("ClassII_TIR", "Transposable element", COUNT_length$Type)
COUNT_length$Type <- gsub("ClassII_Helitrons", "Transposable element", COUNT_length$Type)
COUNT_length$Type <- gsub("mitochondria", "Mitochondria", COUNT_length$Type)
COUNT_length$Type <- gsub("chloroplast", "Plastid", COUNT_length$Type)
COUNT_length$Type <- gsub("repeat_region", "Other repeat", COUNT_length$Type)
COUNT_length$Type <- gsub("telomere", "Telomere", COUNT_length$Type)
COUNT_length$Type <- gsub("Nonrepetitive_genome", "Non-repetitive genome", COUNT_length$Type)


# Define colours
ELEMENTS <- c("Non-repetitive genome",  "Mitochondria", "Plastid", "Telomere", "45S rDNA", "5S rDNA", "Centromere", "Other repeat","Transposable element")
myColors_UNPLACED <-  c('#1a9846', "#8c510a", "#dfc27d", '#8856a7', '#ce1256', '#c994c7', '#67001f', '#bdbdbd', '#3690c0')
names(myColors_UNPLACED) <- as.factor(ELEMENTS)

# Re-order repeats 
COUNT_length$Type <- factor(COUNT_length$Type, levels=c(ELEMENTS), exclude=NULL)   

# Set 'theme_minimal' as standard theme 
theme_set(  theme_minimal() )

# Generate PLOT
plot_scaff_vs_unplaced <- ggplot(COUNT_length, aes(fill=Type, y=Accession, x=Length/1000000)) + 
  geom_bar(stat = "identity", position = "stack", width = 0.85) +
  scale_fill_manual("Annotation",  values = myColors_UNPLACED,  na.value="#969696" ) +
  labs(x = "Cumulative length (Mb)", y = "Accession", title = "") +
  theme(axis.text.x = element_text(angle = 0, vjust = 1, hjust=0.5)) +
  theme(legend.position = "bottom", text = element_text(size=18), axis.text.y = element_text(angle = 0, vjust = 0.5, hjust=1, size=15), axis.text.x = element_text(size=15)) +
  facet_grid(. ~ Source, scales="free_x") +
  theme(strip.background = element_rect( color="#bababa", fill="#bababa", linetype="solid"), strip.text.x = element_text(size = 18, color = "black"), panel.spacing = unit(1, "cm", data = NULL)) +
  guides(fill=guide_legend(ncol=3))

##face = "bold.italic"
##  ggtitle(label = "Contributions of repeats\nSCAFFOLDED vs UNPLACED") +

print(plot_scaff_vs_unplaced)

# PDF
#pdf(paste0('/Users/frabanal/Downloads/SuppFig.Repeats_scaffolded_vs_unplaced.pdf'), width = 10, height = 8)
pdf(paste0(workdir, '/4_plots/SuppFig.Repeats_scaffolded_vs_unplaced.pdf'), width = 10, height = 8)
plot_scaff_vs_unplaced
dev.off()

# PNG
#png(paste0('/Users/frabanal/Downloads/SuppFig.Repeats_scaffolded_vs_unplaced.png'), width = 250, height = 200, units='mm', res = 300)
png(paste0(workdir, '/4_plots/SuppFig.Repeats_scaffolded_vs_unplaced.png'), width = 250, height = 200, units='mm', res = 300)
plot_scaff_vs_unplaced
dev.off()

print(plot_scaff_vs_unplaced)


```
Plot Total length distribution of TEs

```{r}
head(COUNT_length_PLACED_)

# Subset TEs
TE_classes <- subset(COUNT_length_PLACED_, Type2 == 'ClassI_LTR' |  Type2 == 'ClassI_nonLTR' | Type2 == 'ClassII_Helitrons' | Type2 == 'ClassII_TIR') 

# Substitute name
TE_classes$Type2 <- gsub("ClassI_LTR", "Class I LTR", TE_classes$Type2)
TE_classes$Type2 <- gsub("ClassI_nonLTR", "Class I non-LTR", TE_classes$Type2)
TE_classes$Type2 <- gsub("ClassII_TIR", "Class II TIR", TE_classes$Type2)
TE_classes$Type2 <- gsub("ClassII_Helitrons", "Class II Helitron", TE_classes$Type2)

# Define colours
ELEMENTS_ <- c("Class I LTR", "Class I non-LTR", "Class II TIR", "Class II Helitron")
myColors_TEs_ <-  c('#023858', '#0570b0', "#d0d1e6", "#a6bddb")
names(myColors_TEs_) <- as.factor(ELEMENTS_)
# Average color: #3690c0


plot_TEs <- ggplot(TE_classes, aes(x=Length/1000000, fill=Type2)) +
    geom_histogram(alpha=0.7, position="identity", aes(y = ..density..), color="black") +
    geom_density(alpha=0.7) +
    scale_fill_manual("Annotation", values = myColors_TEs_,  na.value="#d4b9da") +
    theme_minimal() +
    theme(legend.position = c(0.84,0.77), text = element_text(size=18), axis.text.y = element_text(angle = 0, vjust = 0.5, hjust=1, size=15), axis.text.x = element_text(size=15)) +
    labs(x="Cumulative length (Mb)", y = "Density") 

print(plot_TEs)

# PDF
#pdf(paste0('/Users/frabanal/Downloads/MSuppFig.TEs.pdf'), width = 8, height = 4)
pdf(paste0(workdir, '/4_plots/SuppFig.TEs.pdf'), width = 8, height = 4)
plot_TEs
dev.off()

# PNG
#png(paste0('/Users/frabanal/Downloads/MainFig.Repeats_blast.png'), width = 80, height = 200, units='mm', res = 300)
png(paste0(workdir, '/4_plots/SuppFig.TEs.png'), width = 200, height = 100, units='mm', res = 300)
plot_TEs
dev.off()

print(plot_TEs)


```








































