---
title: "Compare annotations"
output: null_document
---

# Setup
```{r}
library(pannagram)
library(crayon)
library(rhdf5)
library(ggplot2)
```

## Paths
```{r}
path.msa = '/Volumes/Samsung_T5/vienn/test/a27/intermediate/consensus2/'
path.msa = '/Volumes/Samsung_T5/vienn/test/a27/annot5/'
path.prev = '/Users/annaigolkina/Library/CloudStorage/OneDrive-Personal/vienn/pacbio/1001Gplus/01_data/04_annotation/02_pannagram/genes_v05/'
```


```{r}
# ---- Manual testing ----


gff.new.own = read.table(paste0(path.msa, 'gff_own.gff'), 
                         stringsAsFactors = F)
gff.new.pan = read.table(paste0(path.msa,'gff_pan.gff'),
                         stringsAsFactors = F)

accessions = unique(gff.new.own$V1)
accessions = sapply(accessions, function(s) strsplit(s, '_')[[1]][1])
accessions = unique(accessions)
length(accessions)


for(i.chr in 1:5){
  s.prev = paste0('AT', i.chr, "Gr")
  s.new = paste0('AT', i.chr, "SG")
  gff.new.own$V9 = gsub(s.prev, s.new, gff.new.own$V9)
}


gr.table.all = c()
for(acc in accessions){
  pokaz('Accession', acc)
  x.init = read.table(paste0(path.prev, 'genes_v05_',acc,'.gff'))
  y.init = gff.new.own[grep(paste0(acc, '_Chr'), gff.new.own$V1),]
  
  y = y.init[y.init$V3 == 'mRNA',]
  x = x.init[x.init$V3 == 'mRNA',]
  pokaz(acc, nrow(x), nrow(y))
  
  x$cover = 0
  x$id = 1:nrow(x)
  
  y$V10 = ''
  y$id = 1:nrow(y)
  
  i.chr = 1
  s.strand = '+'
  for(i.chr in 1:5){
    for(s.strand in c('+', '-')){
      
      pokaz(i.chr, s.strand)
      
      xi = x[x$V1 == paste0(acc, '_Chr', i.chr),]
      yi = y[y$V1 == paste0(acc, '_Chr', i.chr),]
      xi = xi[xi$V7 == s.strand,]
      yi = yi[yi$V7 == s.strand,]
      
      pos.x = rep(0, 40000000)
      idx = c()
      for(irow in 1:nrow(xi)){
        pos.x[xi$V4[irow]:xi$V5[irow]] = irow
      }
      
      for(irow in 1:nrow(yi)){
        pp = pos.x[yi$V4[irow]:yi$V5[irow]]
        m = mean(pp > 0)
        if(m < 0.5) next
        p = unique(pp)
        p = setdiff(p, 0)
        if(length(p) == 1){
          if(y[yi$id[irow],]$V10 != '') stop('1')
          y[yi$id[irow],]$V10 = xi$V9[p]
        } else if (length(p) > 1) {
          if(y[yi$id[irow],]$V10 != '') stop('2')
          y[yi$id[irow],]$V10 = paste(xi$V9[p], collapse = '|')
        }
      }
      
      # Coverage of X-genes
      pos.y = rep(0, 40000000)
      idx = c()
      for(irow in 1:nrow(yi)){
        pos.y[yi$V4[irow]:yi$V5[irow]] = irow
      }
      
      idx = c()  # IDXs of genes, which were not found in the second iteration
      for(irow in 1:nrow(xi)){
        m = mean(pos.y[xi$V4[irow]:xi$V5[irow]])
        x$cover[xi$id[irow]] = m
      }
    }
  }
  
  gr.table = y[,c('V9', 'V10')]
  idx.twogenes <- grep("\\|", gr.table$V10)
  gr.table = gr.table[-idx.twogenes,]
  gr.table = gr.table[gr.table$V10 != '',]
  
  for(icol in 1:ncol(gr.table)){
    gr.table[,icol] = sapply(gr.table[,icol], function(s) strsplit(s, ';')[[1]][1])
    gr.table[,icol] = sapply(gr.table[,icol], function(s) strsplit(s, '\\.')[[1]][1])
    gr.table[,icol] = sapply(gr.table[,icol], function(s) strsplit(s, '=')[[1]][2])
  }
  if(sum(is.na(gr.table)) > 0) stop('NA detected')

  # Remove dupliced genes  
  gf.dup = gr.table$V10[duplicated(gr.table$V10)]
  gr.table = gr.table[!(gr.table$V10 %in% gf.dup),]
  gr.table$acc = acc
  
  gr.table.all = rbind(gr.table.all, gr.table)
  
  # for(irow in 1:nrow(gr.table)){
  #   y.init$V9 = gsub(gr.table$V9[irow], gr.table$V10[irow], y.init$V9)
  # }
  # 
  # library(stringi)
  # 
  # y.init$V9 <- stri_replace_all_fixed(
  #   y.init$V9,
  #   pattern = gr.table$V9,
  #   replacement = gr.table$V10,
  #   vectorize_all = FALSE
  # )
  
}
```

## Save
```{r}
saveRDS(gr.table.all, paste0(path.msa,'gr_table_all.rds'))
```

## Read
```{r}
gr.table.all = readRDS(paste0(path.msa,'gr_table_all.rds'))

```

# Duplicated
```{r}
gr.table1 = unique(gr.table.all[,1:2])

gr.dup9 = gr.table1[duplicated(gr.table1[,2]),2]

length(unique(gr.table.all$V10))

gr.dup9 = sort(gr.dup9)

```

## Example
```{r}

gr.dup = gr.dup9

i.g = 2

g = gr.dup[i.g]
gr2 = sort(unique(gr.table1[gr.table1[,2] == g,1]))

groups.target = gsub('SG', 'Gr', gr2)

tmp = c()
for(g in groups.target){
  pokaz(g)
  idx = grep(g, gff.new.pan$V9)
  tmp = rbind(tmp, gff.new.pan[idx,])
}

pos.shift = 1000
# pos.shift = 0
pos.beg = min(tmp$V4) - pos.shift
pos.end = max(tmp$V5) + pos.shift

pokaz('Length', pos.end - pos.beg)

aln = c()
i.chr = 1
file.comb = paste0(path.msa, '/extra2_1_1.h5')
path.chr = '/Volumes/Samsung_T5/vienn/genomes/pb_chromosomes/'
gr.accs.e <- "accs/"

accessions.true = c('0', "10002", "10015", "10024", "1741", "220011", "22002", "22003", "22004", 
                    "22005", "22006", "22007", "6024", "6069", "6124", "6244", "6909", 
                    "6966", "8236", "9075", "9537", "9543", "9638", "9728", "9764", 
                    "9888", "9905", "9981")

for(acc in accessions.true){
  
  pokaz('Accession', acc)
  
  if(acc == '220011'){
    v = h5read(file.comb, paste0(gr.accs.e, '22001_mod'))
  } else {
    v = h5read(file.comb, paste0(gr.accs.e, acc))  
  }
  
  v.pos = v[pos.beg:pos.end]
  
  file.chr = paste0(path.chr, acc, '_chr', i.chr, '.fasta')  
  s.chr = seq2nt(readFasta(file.chr))
  
  s = rep('-', length(v.pos))
  s[v.pos != 0] = s.chr[abs(v.pos[v.pos != 0])]
  if(sum(v.pos < 0) > 0){
    s[v.pos < 0] = justCompl(s[v.pos < 0])  
  }
  
  aln = rbind(aln, s)
}


x = gff.new.pan[(gff.new.pan$V4 >=pos.beg) & 
                  (gff.new.pan$V5 <= pos.end) & 
                  (gff.new.pan$V3 == 'mRNA'),]
x$chr = as.numeric(sapply(x$V1, function(s) strsplit(s, '_Chr')[[1]][2]))
i.chr = 1
x = x[x$chr == i.chr,]
x = data.frame(beg = x$V4, end = x$V5)

x = unique(x)
orfplot(x)

rownames(aln) = accessions.true
p = msaplot(aln) 
# p

p +  annotate("rect", xmin = x$beg - pos.beg, xmax = x$end - pos.beg,
               # ymin = -Inf, ymax = Inf, alpha = 0.2,
              ymin = 0, ymax = 0.5,  alpha = 0.8,
               fill = 'blue')


s = mx2cons(aln)

save(list = 's', file = paste0(path.msa, "tmp_consensus_", i.g, ".RData"))
```


```{r}
library(ggplot2)
library(pannagram)
# load("tmp_consensus_2.RData")
p = dotplot(s, s, 15, 15)

p

# png(paste('tmp2.png', sep = ''), 
#     width = 8, height = 8, units = "in", res = 300)
# print(p)    
# dev.off()



```


```{r}

s1 = seq2clean(nt2seq(aln['10002',]))
s2 = seq2clean(nt2seq(aln['0',]))


s2 = substr(s2, nchar(s2) - 7999, nchar(s2))


dotplot.s(s1, s2, 15, 13)
dotplot.s(s1, s1, 15, 13)


```



```{r}
gff.new.own[(gff.new.own$V1 == '8236_Chr1') & (grepl('AT1Gr10003195', gff.new.own$V9)),]
```



```{r}

aln.type = 'clean_'
aln.type = 'add_'
aln.type = 'extra1_'
aln.type = 'extra2_'
i.chr = 1
file.comb = paste0('/Volumes/Samsung_T5/vienn/test/a27/annot3/', aln.type, i.chr, '_', i.chr, '.h5')
acc = '220011'

v = h5read(file.comb, paste0(gr.accs.e, acc))  


pos = 11378247:11394180
sum(v[pos] != 0)
min(v)
max(v)

pos = 11983955:11986558
sum(pos %in% v)

```


```{r}
gff.new.own[(gff.new.own$V1 == '1741_Chr1') & (grepl('AT1Gr10001936', gff.new.own$V9)),]
```

# Which alignment has it
```{r}

aln.type = 'add_'
aln.type = 'clean_'
file.comb = paste0('/Volumes/Samsung_T5/vienn/test/a27/annot5/', aln.type, '1_1.h5')
acc = '22005'

v = h5read(file.comb, paste0(gr.accs.e, acc))  

pos = 4383655:4389097

sum(pos %in% v)

findRuns(which(pos %in% v))

pos[c(117,1346)]

```


# MSA of the insertion
```{r}

groups.target = c('AT1Gr10003194', 'AT1Gr10003195', 'AT1Gr10003196')

groups.target = c('AT1Gr10001936', 'AT1Gr10001937')

tmp = c()
for(g in groups.target){
  pokaz(g)
  idx = grep(g, gff.new.pan$V9)
  tmp = rbind(tmp, gff.new.pan[idx,])
}

pos.beg = min(tmp$V4)
pos.end = max(tmp$V5)

pos.shift = 1000
pos.shift = 0

pos.beg = 16724004 - pos.shift
pos.end = 16729172 + pos.shift

aln = c()
i.chr = 1
file.comb = '/Volumes/Samsung_T5/vienn/test/a27/annot3/extra2_1_1.h5'
path.chr = '/Volumes/Samsung_T5/vienn/genomes/pb_chromosomes/'
gr.accs.e <- "accs/"

accessions.true = c('0', "10002", "10015", "10024", "1741", "220011", "22002", "22003", "22004", 
                    "22005", "22006", "22007", "6024", "6069", "6124", "6244", "6909", 
                    "6966", "8236", "9075", "9537", "9543", "9638", "9728", "9764", 
                    "9888", "9905", "9981")

for(acc in accessions.true){
  
  pokaz('Accession', acc)
  
  if(acc == '220011'){
    v = h5read(file.comb, paste0(gr.accs.e, '22001_mod'))
  } else {
    v = h5read(file.comb, paste0(gr.accs.e, acc))  
  }
  
  v.pos = v[pos.beg:pos.end]
  
  file.chr = paste0(path.chr, acc, '_chr', i.chr, '.fasta')  
  s.chr = seq2nt(readFasta(file.chr))
  
  s = rep('-', length(v.pos))
  s[v.pos != 0] = s.chr[abs(v.pos[v.pos != 0])]
  if(sum(v.pos < 0) > 0){
    s[v.pos < 0] = justCompl(s[v.pos < 0])  
  }
  
  aln = rbind(aln, s)
}



x = data.frame(beg = tmp$V4, end = tmp$V5)
orfplot(x)

rownames(aln) = accessions.true
msaplot(aln) +  annotate("rect", xmin = x$beg - min(x$beg), xmax = x$end -  min(x$beg),
               ymin = -Inf, ymax = Inf, alpha = 0.2, fill = 'blue')



```


pos = 11983955:11986558

# Test where the insertion happend
```{r}

aln.type = 'clean_'
aln.type = 'msa_'
aln.type = 'extra1_'
aln.type = 'extra2_'

# Positions
pos = 16724004:16729172; acc = '10002'; i.chr = 1
pos = 32093916:32102352; acc = '10002'; i.chr = 1
pos = 20365527:20370142; acc = '10002'; i.chr = 2
pos = 17726:20862; acc = '10024'; i.chr = 1
pos = 1794442:1802750; acc = '22005'; i.chr = 1

file.comb = paste0('/Volumes/Samsung_T5/vienn/test/a27/annot3/', aln.type, i.chr, '_', i.chr, '.h5')

pos = 5249458:5258408; acc = '22005'; i.chr = 1
file.comb = paste0('/Volumes/Samsung_T5/vienn/test/a27/annot5/', aln.type, i.chr, '_', i.chr, '.h5')

v = h5read(file.comb, paste0(gr.accs.e, acc))  

sum(v[pos] != 0)
tmp = v[pos]
tmp = tmp[tmp !=0]
p1 = min(tmp)
p2 = max(tmp)

pokaz(p1, p2, p2 - p1 +1)
# p1 = 11983955
# p2 = 11986558


```


## New alignment
```{r}

aln.type = 'msa_'
file.comb = paste0('/Volumes/Samsung_T5/vienn/test/a27/annot5/', aln.type, i.chr, '_', i.chr, '.h5')

v = h5read(file.comb, paste0(gr.accs.e, acc))  

pos = p1:p2

sum(pos %in% v)

idx = which(v %in% pos)

max(idx)

```

## MSA in new aln
```{r}
pos.shift = 1000
# pos.shift = 0

pos.beg = min(idx) - pos.shift
pos.end = max(idx) + pos.shift

path.chr = '/Volumes/Samsung_T5/vienn/genomes/pb_chromosomes/'
gr.accs.e <- "accs/"

accessions.true = c('0', "10002", "10015", "10024", "1741", "220011", "22002", "22003", "22004", 
                    "22005", "22006", "22007", "6024", "6069", "6124", "6244", "6909", 
                    "6966", "8236", "9075", "9537", "9543", "9638", "9728", "9764", 
                    "9888", "9905", "9981")

aln = c()
for(acc in accessions.true){
  
  pokaz('Accession', acc)
  
  if(acc == '220011'){
    v = h5read(file.comb, paste0(gr.accs.e, '22001_mod'))
  } else {
    v = h5read(file.comb, paste0(gr.accs.e, acc))  
  }
  
  v.pos = v[pos.beg:pos.end]
  
  file.chr = paste0(path.chr, acc, '_chr', i.chr, '.fasta')  
  s.chr = seq2nt(readFasta(file.chr))
  
  s = rep('-', length(v.pos))
  s[v.pos != 0] = s.chr[abs(v.pos[v.pos != 0])]
  if(sum(v.pos < 0) > 0){
    s[v.pos < 0] = justCompl(s[v.pos < 0])  
  }
  
  aln = rbind(aln, s)
}

rownames(aln) = accessions.true
msaplot(aln) 

```


# Create a test dataset
```{r}

path.test = '/Volumes/Samsung_T5/vienn/test/a27/test/'

aln.type = 'ref_'
# Find which ids to take
idx = list()
i.chr = 1
acc = '22005'
pos = 4383771:4385000
for(ref.name in c('0', '10015', '10024')){
  file.comb = paste0('/Volumes/Samsung_T5/vienn/test/a27/annot5/', aln.type, i.chr, '_', i.chr,'_', ref.name, '.h5')

  v = h5read(file.comb, paste0(gr.accs.e, acc))   
  
  tmp = which(v %in% pos)
  idx[[ref.name]] = min(tmp):max(tmp)
}



vv = list()
for(ref.name in c('0', '10015', '10024')){
  file.comb = paste0('/Volumes/Samsung_T5/vienn/test/a27/annot5/', aln.type, i.chr, '_', i.chr,'_', ref.name, '.h5')
  
  v.mx = c()
  for(acc in c('0', '10015', '10024', '22005', '22006')){
    pokaz(ref.name, acc)
    v = h5read(file.comb, paste0(gr.accs.e, acc))    
    v = v[idx[[ref.name]]]
    v.mx = cbind(v.mx, v)
  }
  colnames(v.mx) = c('0', '10015', '10024', '22005', '22006')
  vv[[ref.name]]  = v.mx
}


for(ref.name in c('0', '10002', '10015')){
  pokaz('Ref', ref.name)
  file.comb = paste0('/Volumes/Samsung_T5/vienn/test/a27/annot5/', aln.type, i.chr, '_', i.chr,'_', ref.name, '.h5')
  file.comb.out = paste0(path.test, aln.type, i.chr, '_', i.chr,'_', ref.name, '.h5')
  
  h5createFile(file.comb.out)
  h5createGroup(file.comb.out, gr.accs.e)
  
  groups1 = h5ls(file.comb)
  
  accessions = groups1$name[groups1$group == gr.accs.b]
  
  for(acc in accessions){
    v = h5read(file.comb, paste0(gr.accs.e, acc))   
    v = v[idx[[ref.name]]]
    
    suppressMessages({
      s = paste0('/',gr.accs.e, acc)
      h5write(v, file.comb.out, s)
    })
  }
      
}







```



```{r}


aln.type = 'ref_'
i.chr = 1

pos = 11983955:11986558
pos.common = pos
for(ref.name in c('0', '10002', '10015')){
  file.comb = paste0('/Volumes/Samsung_T5/vienn/test/a27/annot3/', aln.type, i.chr, '_', i.chr,'_', ref.name, '.h5')
  acc = '10002'
  
  v = h5read(file.comb, paste0(gr.accs.e, acc))   
  
  pp = which(v %in% pos)
  
  pos.common = intersect(pos.common, v)
  
  pokaz(length(pos.common))
}


# Find which ids to take
idx = list()
for(ref.name in c('0', '10002', '10015')){
  file.comb = paste0('/Volumes/Samsung_T5/vienn/test/a27/annot3/', aln.type, i.chr, '_', i.chr,'_', ref.name, '.h5')
  acc = '10002'
  
  v = h5read(file.comb, paste0(gr.accs.e, acc))   
  
  idx[[ref.name]] = which(v %in% pos.common)
}



vv = list()
for(ref.name in c('0', '10002', '10015')){
  file.comb = paste0('/Volumes/Samsung_T5/vienn/test/a27/annot3/', aln.type, i.chr, '_', i.chr,'_', ref.name, '.h5')
  
  v.mx = c()
  for(acc in accessions.true){
    if(acc == '220011') acc = '22001_mod'
    v = h5read(file.comb, paste0(gr.accs.e, acc))    
    v = v[idx[[ref.name]]]
    v.mx = cbind(v.mx, v)
  }
  colnames(v.mx) = accessions.true
  vv[[ref.name]]  = v.mx
}



```


```{r}
i = 1
j = 2
k = 3

vi = vv[[i]]
vj = vv[[j]]
vk = vv[[k]]

vi[vj == 0] = 0
vj[vi == 0] = 0

sum(rowSums(vi != vj) == 0)


```


```{r}
x = breaks[(breaks$idx.beg >= 11546625) & 
         (breaks$idx.end <= 11548626) ,]

x = x[order(-x$len.acc),]

```


# Check what other genomes show in the clean for positions 12076719 - 12077176
```{r}

aln.type = 'clean_'
file.comb = paste0('/Volumes/Samsung_T5/vienn/test/a27/annot4/', aln.type, i.chr, '_', i.chr, '.h5')

file.comb = '/Volumes/Samsung_T5/vienn/test/a27/annot4/comb_1_1.h5'

# pos.target = 11547625:11548406

pos.target = 12076718:12077176

v.mx = c()
for(acc in accessions.true){
  pokaz(acc)
  if(acc == '220011') acc = '22001_mod'
  v = h5read(file.comb, paste0(gr.accs.e, acc))    
  v = v[pos.target]
  v.mx = cbind(v.mx, v)
}

colnames(v.mx) = accessions.true[1:ncol(v.mx)]

```






# Example
```{r}

s = readFasta('/Users/annaigolkina/Library/CloudStorage/OneDrive-Personal/vienn/pacbio/1001Gplus/02_analysis_git/03_annotation/02_scripts/figures_compare_good/group_2_14054188_14076618.fasta')

s = seq2clean(s)

dotplot.s(s['10015'], s['10024'], 15, 12)



```

## MSA plot
```{r}

# Example 1
i.chr = 2
pos.shift = 1000
pos.beg = 14054188 - pos.shift
pos.end = 14076618 + pos.shift

# Example 2
i.chr = 1
pos.shift = 1000
pos.beg = 4489640 - pos.shift - 10000
pos.end = 4497090 + pos.shift

#msaplot_1_26677934_26708883
# Example 3
i.chr = 1
pos.shift = 1000
pos.beg = 12151829 - pos.shift
pos.end = 12188410 + pos.shift

# Example 3.2
i.chr = 1
pos.shift = 1000
pos.beg = 12052416 - pos.shift
pos.end = 12056127 + pos.shift


file.comb = paste0('/Volumes/Samsung_T5/vienn/test/a27/annot5/extra2_',i.chr,'_',i.chr,'.h5')
path.chr = '/Volumes/Samsung_T5/vienn/genomes/pb_chromosomes/'
gr.accs.e <- "accs/"

accessions.true = c('0', "10002", "10015", "10024", "1741", "220011", "22002", "22003", "22004", 
                    "22005", "22006", "22007", "6024", "6069", "6124", "6244", "6909", 
                    "6966", "8236", "9075", "9537", "9543", "9638", "9728", "9764", 
                    "9888", "9905", "9981")

aln = c()
for(acc in accessions.true){
  
  pokaz('Accession', acc)
  
  if(acc == '220011'){
    v = h5read(file.comb, paste0(gr.accs.e, '22001_mod'))
  } else {
    v = h5read(file.comb, paste0(gr.accs.e, acc))  
  }
  
  v.pos = v[pos.beg:pos.end]
  
  file.chr = paste0(path.chr, acc, '_chr', i.chr, '.fasta')  
  s.chr = seq2nt(readFasta(file.chr))
  
  s = rep('-', length(v.pos))
  s[v.pos != 0] = s.chr[abs(v.pos[v.pos != 0])]
  if(sum(v.pos < 0) > 0){
    s[v.pos < 0] = justCompl(s[v.pos < 0])  
  }
  
  aln = rbind(aln, s)
}


rownames(aln) = accessions.true
msaplot(aln) 


```

```{r}

s = mx2seq(aln)

acc1 = '0'
acc2 = '220011'
p = dotplot.s(s[acc1], s[acc2], 15, 12)

p + xlab(acc1) + ylab(acc2)

```

## Before

```{r}

aln.type = 'extra2_'

# Positions
pos = pos.beg:pos.end; acc = '0';

# Alignment with confusing positions
file.comb = paste0('/Volumes/Samsung_T5/vienn/test/a27/annot5/', aln.type, i.chr, '_', i.chr, '.h5')
v = h5read(file.comb, paste0(gr.accs.e, acc))  


sum(v[pos] != 0)
tmp = v[pos]
tmp = tmp[tmp !=0]
p1 = min(tmp)
p2 = max(tmp)


pokaz(p1, p2, p2 - p1 +1)
# p1 = 11983955
# p2 = 11986558


```


### Prev alignment
```{r}

aln.type = 'v_'
file.comb = paste0('/Users/annaigolkina/Library/CloudStorage/OneDrive-Personal/vienn/pacbio/1001Gplus/01_data/02_alignment/pannagram_v05/', aln.type, i.chr, '_', i.chr, '_ref_0.h5')

v = h5read(file.comb, paste0(gr.accs.e, acc))  

pos = p1:p2

sum(pos %in% v)

idx = which(v %in% pos)

max(idx)

```

## MSA in new aln
```{r}
#pos.shift = 1000
pos.shift = 0

pos.beg = min(idx) - pos.shift
pos.end = max(idx) + pos.shift

path.chr = '/Volumes/Samsung_T5/vienn/genomes/pb_chromosomes/'
gr.accs.e <- "accs/"

accessions.true = c('0', "10002", "10015", "10024", "1741", "220011", "22002", "22003", "22004", 
                    "22005", "22006", "22007", "6024", "6069", "6124", "6244", "6909", 
                    "6966", "8236", "9075", "9537", "9543", "9638", "9728", "9764", 
                    "9888", "9905", "9981")

aln = c()
for(acc in accessions.true){
  
  pokaz('Accession', acc)
  
  v = h5read(file.comb, paste0(gr.accs.e, acc))  
  v.pos = v[pos.beg:pos.end]
  v.pos[is.na(v.pos)] = 0
  
  file.chr = paste0(path.chr, acc, '_chr', i.chr, '.fasta')  
  s.chr = seq2nt(readFasta(file.chr))
  
  s = rep('-', length(v.pos))
  s[v.pos != 0] = s.chr[abs(v.pos[v.pos != 0])]
  if(sum(v.pos < 0) > 0){
    s[v.pos < 0] = justCompl(s[v.pos < 0])  
  }
  
  aln = rbind(aln, s)
  
  # positions in accessions  
  v.pos = v.pos[v.pos != 0]
  if(length(v.pos) == 0){
    pokaz(acc, "Empty")
  } else {
    pokaz(acc, min(v.pos), max(v.pos))  
  }
  

  
}

rownames(aln) = accessions.true[1:nrow(aln)]
msaplot(aln) 

msadiff(aln) 

```

### Lost positions
```{r}

acc = '6244'
pos.lost = 9412573:9416231


aln.type = 'extra2_'

# Alignment with confusing positions
file.comb = paste0('/Volumes/Samsung_T5/vienn/test/a27/annot5/', aln.type, i.chr, '_', i.chr, '.h5')
v = h5read(file.comb, paste0(gr.accs.e, acc))  

tmp = which(v %in% pos.lost)
p1 = min(tmp)
p2 = max(tmp)

pokaz(p1, p2, p2 - p1 +1)

pokaz(which(v == p1), which(v == p2))



```




# Comparrison
```{r}
library(igraph)
```

## Read annotations
```{r}
path.msa = '/Volumes/Samsung_T5/vienn/test/a27/annot5/'
gff.new = read.table(paste0(path.msa,'gff_pan.gff'),
                         stringsAsFactors = F)

gff.pre = read.table('/Users/annaigolkina/Library/CloudStorage/OneDrive-Personal/vienn/pacbio/1001Gplus/01_data/04_annotation/02_pannagram/genes_v05/genes_v05_pangen_all.gff', stringsAsFactors = F)

gff.new = gff.new[gff.new$V3 == 'mRNA',]
gff.pre = gff.pre[gff.pre$V3 == 'gene',]

gff.new$gr = sapply(gff.new$V9, function(s) strsplit(s, '\\.')[[1]][1])


# matrix with positions on groups
pos.new = data.frame(beg = tapply(gff.new$V4, gff.new$gr, min),
                     end = tapply(gff.new$V5, gff.new$gr, max))

pos.new$gr = rownames(pos.new)
pos.new$gr = sub('ID=', '', pos.new$gr)
pos.new$gr = sub('Gr', 'SG', pos.new$gr)
rownames(pos.new) = pos.new$gr


pos.pre = data.frame(beg = gff.pre$V4,
                     end = gff.pre$V5,
                     gr = sapply(gff.pre$V9, function(s) strsplit(s, ';')[[1]][1]))
pos.pre = pos.pre[nchar(pos.pre$gr) == 16,]
pos.pre$gr = sub('ID=', '', pos.pre$gr)
rownames(pos.pre) = pos.pre$gr

```


## Duplicated
```{r}

gr.table.all = readRDS('/Volumes/Samsung_T5/vienn/test/a27/annot5/gr_table_all.rds')

gr.table.all = gr.table.all[,-3]
gr.table.all = unique(gr.table.all)

dim(gr.table.all)

gr.dup1 = unique(gr.table.all[duplicated(gr.table.all[,1]),1])
gr.dup2 = unique(gr.table.all[duplicated(gr.table.all[,2]),2])
pokaz(length(gr.dup1), length(gr.dup2))

gr.comp <- getGraphComponents(gr.table.all)

pokaz('Total number of super-groups', gr.comp$no)
pokaz('Number of one2one correspondence is', sum(gr.comp$csize == 2))
pokaz('Number of confusing loci is', gr.comp$no - sum(gr.comp$csize == 2))


pokaz("Unique genes new:", length(setdiff(pos.new$gr, gr.table.all$V9)))

pokaz("Unique genes new:", length(setdiff(pos.pre$gr, gr.table.all$V10)))

```


### Examples

```{r}

gr.super1 = c('AT1SG10000357',  'AT1SG10000358' ,'AT1SG10000359')
gr.super2 = gr.table.all[gr.table.all[,1] %in% gr.super1, 2]


gr.super1 = c("AT3SG20000551","AT3SG20000552" ,"AT3SG20000553", "AT3SG20000554" ,"AT3SG20000555", "AT3SG20000556", "AT3SG20000557" ,"AT3SG20000558", "AT3SG20000559" ,"AT3SG20000560" ,"AT3SG20000561" ,"AT3SG20000562", "AT3SG20000563")
gr.super2 = gr.table.all[gr.table.all[,1] %in% gr.super1, 2]


gr.super2 = c("AT5Gr10001858", "AT5Gr10001859", "AT5Gr10001860", "AT5Gr10001861")
gr.super1 = gr.table.all[gr.table.all[,2] %in% gr.super2, 1]


```



```{r}
cl.super = which(gr.comp$csize != 2)
i.super = 4

cl.id = cl.super[i.super]
# cl.id = 12375

gr.super.all = names(gr.comp$membership)[gr.comp$membership == cl.id]

gr.super1 = gr.super.all[grep("SG", gr.super.all)]
gr.super2 = gr.super.all[grep("Gr", gr.super.all)]
```


```{r}
pokaz('Number before/after', length(gr.super2), length(gr.super1))

# Find positions

pos1 = pos.new[gr.super1,c('beg', 'end')]
pos2 = pos.pre[gr.super2,c('beg', 'end')]

pos.aln1 = c(min(pos1$beg), max(pos1$end))
pos.aln2 = c(min(pos2$beg), max(pos2$end))



```

### MSA1

```{r}

i.chr = as.numeric(seq2nt(gr.super1[1])[3])
pos.shift = 1000
pos.beg = pos.aln1[1] - pos.shift
pos.end = pos.aln1[2] + pos.shift

file.comb = paste0('/Volumes/Samsung_T5/vienn/test/a27/annot5/extra2_',i.chr,'_',i.chr,'.h5')
path.chr = '/Volumes/Samsung_T5/vienn/genomes/pb_chromosomes/'
gr.accs.e <- "accs/"

accessions.true = c('0', "10002", "10015", "10024", "1741", "220011", "22002", "22003", "22004", 
                    "22005", "22006", "22007", "6024", "6069", "6124", "6244", "6909", 
                    "6966", "8236", "9075", "9537", "9543", "9638", "9728", "9764", 
                    "9888", "9905", "9981")

aln = c()
for(acc in accessions.true){
  
  pokaz('Accession', acc)
  
  if(acc == '220011'){
    v = h5read(file.comb, paste0(gr.accs.e, '22001_mod'))
  } else {
    v = h5read(file.comb, paste0(gr.accs.e, acc))  
  }
  
  v.pos = v[pos.beg:pos.end]
  
  file.chr = paste0(path.chr, acc, '_chr', i.chr, '.fasta')  
  s.chr = seq2nt(readFasta(file.chr))
  
  s = rep('-', length(v.pos))
  s[v.pos != 0] = s.chr[abs(v.pos[v.pos != 0])]
  if(sum(v.pos < 0) > 0){
    s[v.pos < 0] = justCompl(s[v.pos < 0])  
  }
  
  # positions in accessions  
  v.pos = v.pos[v.pos != 0]
  if(length(v.pos) == 0){
    pokaz(acc, "Empty")
  } else {
    pokaz(acc, min(v.pos), max(v.pos))  
  }
  
  aln = rbind(aln, s)
}

rownames(aln) = accessions.true
p.msa1 = msaplot(aln) 

p.msa1 +  annotate("rect", xmin = pos1$beg -  pos.aln1[1] + pos.shift, xmax = pos1$end -  pos.aln1[1] + pos.shift,
               # ymin = -Inf, ymax = Inf, alpha = 0.2,
              ymin = 0, ymax = 0.5,
               fill = 'blue')


aln1 = aln
```

```{r}

s = mx2seq(aln1)['10024']

writeFasta(s, '/Volumes/Samsung_T5/vienn/test/a27/simsearch/tmp.fasta')


tmp = read.table('/Volumes/Samsung_T5/vienn/test/a27/simsearch/out/simsearch.0_chr1.blast.tmp', stringsAsFactors = F)
tmp = tmp[tmp$V7 > 100,]

plotSynDot(tmp)

df = data.frame(beg = tmp$V2, end = tmp$V3)
orfplot(df)


```


```{r}

s = mx2seq(aln1)[1]
dotplot.s(s, s, 15, 12)


```


### MSA2
```{r}

pos.beg = pos.aln2[1] - pos.shift
pos.end = pos.aln2[2] + pos.shift

path.chr = '/Volumes/Samsung_T5/vienn/genomes/pb_chromosomes/'
gr.accs.e <- "accs/"

accessions.true = c('0', "10002", "10015", "10024", "1741", "220011", "22002", "22003", "22004", 
                    "22005", "22006", "22007", "6024", "6069", "6124", "6244", "6909", 
                    "6966", "8236", "9075", "9537", "9543", "9638", "9728", "9764", 
                    "9888", "9905", "9981")


aln.type = 'v_'
file.comb = paste0('/Users/annaigolkina/Library/CloudStorage/OneDrive-Personal/vienn/pacbio/1001Gplus/01_data/02_alignment/pannagram_v05/', aln.type, i.chr, '_', i.chr, '_ref_0.h5')


aln = c()
for(acc in accessions.true){
  
  pokaz('Accession', acc)
  
  v = h5read(file.comb, paste0(gr.accs.e, acc))  
  v.pos = v[pos.beg:pos.end]
  v.pos[is.na(v.pos)] = 0
  
  file.chr = paste0(path.chr, acc, '_chr', i.chr, '.fasta')  
  s.chr = seq2nt(readFasta(file.chr))
  
  s = rep('-', length(v.pos))
  s[v.pos != 0] = s.chr[abs(v.pos[v.pos != 0])]
  if(sum(v.pos < 0) > 0){
    s[v.pos < 0] = justCompl(s[v.pos < 0])  
  }
  
  aln = rbind(aln, s)
  
  # positions in accessions  
  v.pos = v.pos[v.pos != 0]
  if(length(v.pos) == 0){
    pokaz(acc, "Empty")
  } else {
    pokaz(acc, min(v.pos), max(v.pos), max(v.pos) - min(v.pos), length(v.pos))  
  }
  
}

rownames(aln) = accessions.true[1:nrow(aln)]
aln2 = aln


p.msa2 = msaplot(aln2) 

# p.msa2 = msadiff(aln2) 

p.msa2 +  annotate("rect", 
                   xmin = pos2$beg -  pos.aln2[1] + pos.shift, 
                   xmax = pos2$end -  pos.aln2[1] + pos.shift,
                   ymin = 0, ymax = 0.5,
                   fill = 'blue')


```

## Split/merge decision?
```{r}
gff.common = read.table('/Volumes/Samsung_T5/vienn/test/a27/annot5/annotation/common.txt', stringsAsFactors = F)
gff.common = gff.common[gff.common$V3 == 'gene',]
```


```{r}
gff.tmp = gff.common[(gff.common$V4 >= pos.aln1[1]) & (gff.common$V5 <= pos.aln1[2]),]
gff.tmp = gff.tmp[gff.tmp$V1 == paste0('Pannagram_Chr', i.chr),]

gff.tmp = data.frame(beg = gff.tmp$V4, end = gff.tmp$V5, acc = gff.tmp$V10, strand = gff.tmp$V7)
gff.tmp$y = nrow(aln1) - match(gff.tmp$acc, rownames(aln1)) + 1


cols = c('red', 'blue')

p.msa1 +  annotate("rect", xmin = pos1$beg -  pos.aln1[1] + pos.shift, xmax = pos1$end -  pos.aln1[1] + pos.shift,

              ymin = 0, ymax = 0.5,
               fill = 'black') + 
            annotate("rect",
                xmin = gff.tmp$beg - pos.aln1[1] + pos.shift,
                xmax = gff.tmp$end - pos.aln1[1] + pos.shift,
                ymin = gff.tmp$y - as.numeric(as.factor(gff.tmp$strand))/5, 
                ymax = gff.tmp$y - as.numeric(as.factor(gff.tmp$strand))/5 + 0.1,
              fill = cols[as.numeric(as.factor(gff.tmp$strand)) ]
            ) 

```




## Number of genes that are new now
```{r}
gr.add1 = setdiff(rownames(pos.new), gr.table.all[,1])

gr.target = gr.add1[1]

pos.new[gr.target,]




```




# Compare the number of annotation groups
```{r}
xx = c()
for(acc in accessions){
  pokaz('Accession', acc)
  x.init = read.table(paste0(path.prev, 'genes_v05_',acc,'.gff'))
  y.init = gff.new.own[grep(paste0(acc, '_Chr'), gff.new.own$V1),]
  
  y = y.init[y.init$V3 == 'mRNA',]
  x = x.init[x.init$V3 == 'mRNA',]
  pokaz(acc, nrow(x), nrow(y))
  xx  = rbind(xx, c(nrow(x), nrow(y)))
}
```





```{r}
x.init = x
y.init = y

i.chr = 2
y = y.init[y.init$V1 == paste0(acc, '_Chr', i.chr),]
x = x.init[x.init$V1 == paste0(acc, '_Chr', i.chr),]
x = x[x$V7 == '+',]
y = y[y$V7 == '+',]
pokaz(acc, nrow(x), nrow(y))


tmp = y[,4:5]
pos.y = rep(0, 40000000)
for(irow in 1:nrow(tmp)){
  pos.y[tmp$V4[irow]:tmp$V5[irow]] = 1
}


tmp = x[,4:5]
pos.x = rep(0, 40000000)
idx = c()
for(irow in 1:nrow(tmp)){
  pos.x[tmp$V4[irow]:tmp$V5[irow]] = 1
  
  m = mean(pos.y[tmp$V4[irow]:tmp$V5[irow]])
  if(m != 1){
    idx = rbind(idx, c(irow, m))  
    # stop()
  }
}


table(pos.x, pos.y)

# Find, where x is not covered
pos.x.d = pos.x
pos.x.d[pos.y > 0] = 0

tmp = x[,4:5]
idx = c()  # IDXs of genes, whoch were not found in the second iteration
for(irow in 1:nrow(tmp)){
  m = mean(pos.x.d[tmp$V4[irow]:tmp$V5[irow]])
  if(m > 0){
    idx = rbind(idx, c(irow, m))
  }
}


# Examples
idx.d = 233

x.d = x[idx.d,]

aln.type = 'clean_'
gr.accs.e = "accs/"
file.msa = paste0(path.msa, aln.type, i.chr, '_', i.chr, '.h5')
v = h5read(file.msa, paste0(gr.accs.e, acc))
v.p = v[(v != 0) & !(is.na(v))]

idx.g = which(abs(v) %in% abs(x.d$V4:x.d$V5))
sum(length(idx.g))


v[idx.g]

# --

p.prev = max(v.p[v.p <x.d$V4])
p.next = min(v.p[v.p >x.d$V5])

which(v == p.prev)
which(v == p.next)


v.diff = diff(v.p)
idx = which(abs(v.diff)!= 1)

v.p[unique(sort(c(idx, idx+1)))]

# --

gff2gff(path.cons = path.msa,
        gff1 = x.d,
        acc1 = acc,
        acc2 = s.pannagram,
        n.chr = 5,
        aln.type = aln.type,
        s.chr = s.chr,
        exact.match = T,
        remain = T)


gff2 <- data.frame(
  V1 = "1741_Chr2",
  V2 = "EVM",
  V3 = "gene",
  V4 = 1081974,
  V5 = 1086755,
  V6 = ".",
  V7 = "+",
  V8 = ".",
  V9 = "ID=evm.TU.1741_Chr2.237;Name=EVM%20prediction%201741_Chr2.237;OG=OG0019376;AT=AT2G03530",
  stringsAsFactors = FALSE
)


gff2gff(path.cons = path.msa,
        gff1 = gff2,
        acc1 = acc,
        acc2 = s.pannagram,
        n.chr = 5,
        aln.type = aln.type,
        s.chr = s.chr,
        exact.match = T,
        remain = T)


```


