---
title: "SNP analysis"
output: null_document
---


# ON SEVER (cmp_all folder)
```{r}

d = c()
for(i.chr in 1:5){
  x = readRDS(paste('seq_aln_chr_',i.chr,'_ref_add.rds', sep = ''))

  
  combinations <- t(combn(1:28, 2))
  
  combinations = combinations[sample(1:nrow(combinations), 30, replace = FALSE),]
  
  
  for (k in 1:nrow(combinations)) {
    i <- combinations[k, 1]
    j <- combinations[k, 2]
      
    nts = c("C", "T", "A", "G")
    idx = (x[,i] %in% nts) & (x[,j] %in% nts)
    tmp = sum(x[idx,i] != x[idx,j])
    print(c(i, j, tmp))
    d = rbind(d, c(i.chr, i, j, tmp))
    
  }
}



```



```{r}

n = 0
for(i.chr in 1:5){
  pokaz(i.chr)
  x = readRDS(paste('/Volumes/Samsung_T5/vienn/alignment/new/alignments_0/1741_',i.chr,'_',i.chr,'_full.rds', sep = ''))
  
  
  for(i in 1:nrow(x)){
    s1 = seq2nt(x$V8[i])
    s2 = seq2nt(x$V9[i])
    n = n + sum( (s1 != s2) & (s1 != '-') & (s2 != '-') )
  }
  
}
  


```



```{r}

i.chr = 1
x = readRDS(paste('/Volumes/Samsung_T5/vienn/alignment/new/alignments_0/1741_',i.chr,'_',i.chr,'_full.rds', sep = ''))


```


```{r}

s1 = seq2nt(readFastaMy('/Volumes/Samsung_T5/vienn/pb_chromosomes/1741_chr1.fasta'))
s2 = seq2nt(readFastaMy('/Volumes/Samsung_T5/vienn/pb_chromosomes/0_chr1.fasta'))

i.inv = 3
beg1 = df.inv.chr$beg2[i.inv]
end1 = df.inv.chr$end2[i.inv]
beg2 = df.inv.chr$beg1[i.inv]
end2 = df.inv.chr$end1[i.inv]

dotplot(s1[beg1:end1], s2[beg2:end2], 15, 12)

idx1 = max(which(x$V2 <= beg1))
idx2 = min(which(x$V2 >= beg2)) 



plotSynteny(x[idx1:(idx2+1),], vlines = c(beg1, end1), hlines = c(beg2,end2))

```





# Setup
## My Libs
```{r}

path.base = '../../../'

path.pannagram = paste(path.base, '../pannagram/', sep = '')

source(paste(path.pannagram, 'sim/sim_func.R', sep = ''))
source(paste(path.pannagram, 'sv/graph_refinement.R', sep = ''))
source(paste(path.pannagram, 'utils/utils.R', sep = ''))
source(paste(path.pannagram, 'visualisation/dotplot.R', sep = ''))
source(paste(path.pannagram, 'visualisation/msaplot.R', sep = ''))
source(paste(path.pannagram, 'visualisation/orfplot.R', sep = ''))
source(paste(path.pannagram, 'visualisation/visualisation.R', sep = ''))
```


## Libs
```{r}
invisible(library(ggplot2))
invisible(library(rhdf5))
```


## Paths & Variables
```{r}

path.analysis = paste(path.base, '02_analysis/06_snps/', sep = '')
path.work = paste(path.analysis, '01_data/', sep = '')
path.figures = paste(path.analysis, '03_figures/', sep = '')

path.svs = paste(path.base, '01_data/04_annotation/02_pannagram/svs/', sep = '')
path.tair = paste(path.base, '01_data/09_tair10/', sep = '')
path.aln = paste(path.base, '01_data/02_alignment/pannagram_v05/', sep = '')

path.aln = paste(path.base, '01_data/02_alignment/pannagram_v06/', sep = '')
path.aln = paste(path.base, '01_data/02_alignment/pannagram_v07_mafft/', sep = '')

gr.accs.e <- "accs/"
gr.accs.b <- "/accs"
gr.break.e = 'break/'
gr.break.b = '/break'

path.genomes = paste(path.base, '01_data/01_assembly/fasta/', sep = '')
path.chromosomes = paste(path.base, '01_data/01_assembly/cnromosomes/', sep = '')



```




# Count SNPs
```{r}

s.aln = 'v_'
s.aln = 'msa_'
s.aln = 'comb_'
s.aln = 'res_'

diff.acc = c()

# Define accession names and combinations to iterate
i.chr = 1
file.comb = paste(path.aln, s.aln,i.chr,'_',i.chr,'_ref_0.h5', sep = '')

groups = h5ls(file.comb)
accessions = groups$name[groups$group == gr.accs.b]

combinations <- t(combn(accessions, 2))

combinations = combinations[sample(1:nrow(combinations), 30, replace = FALSE),]

for (i in 1:nrow(combinations)) {
  acc1 <- combinations[i, 1]
  acc2 <- combinations[i, 2]
  
  pokaz('Accessions', acc1, acc2)
  
  # Read genomes
  seq1 = readFastaMy(paste(path.genomes, acc1, '.fasta', sep = ''))
  seq2 = readFastaMy(paste(path.genomes, acc2, '.fasta', sep = ''))
  
  for(i.chr in 1:5){
    
    file.comb = paste(path.aln, s.aln,i.chr,'_',i.chr,'_ref_0.h5', sep = '')
    
    v1 = h5read(file.comb, paste(gr.accs.e, acc1, sep = ''))
    v2 = h5read(file.comb, paste(gr.accs.e, acc2, sep = ''))
    
    chr1 = seq2nt(seq1[i.chr])
    chr2 = seq2nt(seq2[i.chr])
    
    v = cbind(v1, v2)
    v = v[rowSums(v == 0) == 0,]
    v = v[rowSums(is.na(v)) == 0,]
    dim(v)
    
    s = cbind(chr1[abs(v[,1])], chr2[abs(v[,2])])
    
    if(sum(v[,1] < 0) > 0){
      s[v[,1] < 0 ,1] = justCompl(s[v[,1] < 0 ,1])  
    }
    
    if(sum(v[,2] < 0) > 0){
      s[v[,2] < 0 ,2] = justCompl(s[v[,2] < 0 ,2])
    }
    d = sum(s[,1] != s[,2])
    
    pokaz('Chr', i.chr, d)
    
    diff.acc = rbind(diff.acc, c(acc1, acc2, i.chr, d))
  }
}


(250632 +156716+168130 + 149222+222072) / 575478 * 478049

(213124 + 128259 + 136666)


```


# Compare, which SNP diapeared and is it good or bad
```{r}

i = 1

  acc1 <- combinations[i, 1]
  acc2 <- combinations[i, 2]
  
  pokaz('Accessions', acc1, acc2)
  
  # Read genomes
  seq1 = readFastaMy(paste(path.genomes, acc1, '.fasta', sep = ''))
  seq2 = readFastaMy(paste(path.genomes, acc2, '.fasta', sep = ''))
  
  i.chr = 1
  
  # previous
  s.aln = 'res_'
  file.comb = paste(path.aln, s.aln,i.chr,'_',i.chr,'_ref_0.h5', sep = '')
  
  v1 = h5read(file.comb, paste(gr.accs.e, acc1, sep = ''))
  v2 = h5read(file.comb, paste(gr.accs.e, acc2, sep = ''))
  
  chr1 = seq2nt(seq1[i.chr])
  chr2 = seq2nt(seq2[i.chr])
  
  v = cbind(v1, v2)
  v = v[rowSums(v == 0) == 0,]
  v = v[rowSums(is.na(v)) == 0,]
  dim(v)
  
  s = cbind(chr1[abs(v[,1])], chr2[abs(v[,2])])
  
  if(sum(v[,1] < 0) > 0){
    s[v[,1] < 0 ,1] = justCompl(s[v[,1] < 0 ,1])  
  }
  
  if(sum(v[,2] < 0) > 0){
    s[v[,2] < 0 ,2] = justCompl(s[v[,2] < 0 ,2])
  }
  d = sum(s[,1] != s[,2])
  
  pokaz('Chr', i.chr, d)
  
  s1 = s
  x1 = v

  # new
  s.aln = 'comb_'
  
  file.comb = paste(path.aln, s.aln,i.chr,'_',i.chr,'_ref_0.h5', sep = '')
  
  v1 = h5read(file.comb, paste(gr.accs.e, acc1, sep = ''))
  v2 = h5read(file.comb, paste(gr.accs.e, acc2, sep = ''))
  
  chr1 = seq2nt(seq1[i.chr])
  chr2 = seq2nt(seq2[i.chr])
  
  v = cbind(v1, v2)
  v = v[rowSums(v == 0) == 0,]
  v = v[rowSums(is.na(v)) == 0,]
  dim(v)
  
  s = cbind(chr1[abs(v[,1])], chr2[abs(v[,2])])
  
  if(sum(v[,1] < 0) > 0){
    s[v[,1] < 0 ,1] = justCompl(s[v[,1] < 0 ,1])  
  }
  
  if(sum(v[,2] < 0) > 0){
    s[v[,2] < 0 ,2] = justCompl(s[v[,2] < 0 ,2])
  }
  d = sum(s[,1] != s[,2])
  
  pokaz('Chr', i.chr, d)
  
  s2 = s
  x2 = v


```

# SNPs saturation
```{r}


s.aln = 'v_'
s.aln = 'msa_'
s.aln = 'comb_'
s.aln = 'res_'

diff.acc = c()

# Define accession names and combinations to iterate
i.chr = 1
file.comb = paste(path.aln, s.aln,i.chr,'_',i.chr,'_ref_0.h5', sep = '')

groups = h5ls(file.comb)
accessions = groups$name[groups$group == gr.accs.b]

for(n.accs in 2:26){
  
  
  combinations <- t(combn(accessions, n.accs))
  
  combinations = combinations[sample(1:nrow(combinations), 
                                     min(nrow(combinations),20), 
                                     replace = FALSE),]
  
  for (i.comb in 1:nrow(combinations)) {
    pokaz('Combination', i.comb)
    
    for(i.chr in 1:5){
      file.comb = paste(path.aln, s.aln,i.chr,'_',i.chr,'_ref_0.h5', sep = '')
      
      for(acc in combinations[i.comb,]){
          
        # Read chromosomes
        chr1 = seq2nt(readFastaMy(paste(path.chromosomes, acc1, '_chr', i.chr, '.fasta', sep = '')))
        
        v = h5read(file.comb, paste(gr.accs.e, acc1, sep = ''))
        
        s = rep('-', length(v))
        s[v != 0] = chr2[abs(v)]
        
        if(sum(v[,1] < 0) > 0){
          s[v[,1] < 0 ,1] = justCompl(s[v[,1] < 0 ,1])  
        }
        
        if(sum(v[,2] < 0) > 0){
          s[v[,2] < 0 ,2] = justCompl(s[v[,2] < 0 ,2])
        }
        d = sum(s[,1] != s[,2])
        
        pokaz('Chr', i.chr, d)
        
        diff.acc = rbind(diff.acc, c(acc1, acc2, i.chr, d))
        
        
      } # accession
    } # chromosome
  } # one combination
} # number fo combinations


```



# OLD
```{r}

s.list = list()
for(i.chr in 1:5){
  s.list[[i.chr]] = list()
}

for (acc in accessions) {
  
  pokaz('Accessions', acc)
  
  # Read genomes
  seq = readFastaMy(paste(path.genomes, acc, '.fasta', sep = ''))
  
  for(i.chr in 1:5){
    
    file.comb = paste(path.aln, 'v_',i.chr,'_',i.chr,'_ref_0.h5', sep = '')
    
    v = h5read(file.comb, paste(gr.accs.e, acc, sep = ''))
    
    s.chr = seq2nt(seq[i.chr])
    
    s = rep('-', length(v))
    idx.ok = which((!is.na(v)) & (v != 0))
    s[idx.ok] = s.chr[abs(v[idx.ok])]
    
    s.list[[i.chr]][[acc]] = s
    
  }
}



```



