---
title: "Translate all raw gene annotation gff files for pangenome coordicates"
output: null_document
---


# Setup
```{r}
library(rhdf5)
source('../../../../pannagram/analys/analys_func.R')
source('../../../../pannagram/utils/utils.R')
source('../../../../pannagram/pangen/comb_func.R')

path.raw = '../../../01_data/04_annotation/01_raw_max/'
path.pan.raw = '../../../01_data/04_annotation/02_pannagram/genes_v05_raw/'
path.pan.genes = '../../../01_data/04_annotation/02_pannagram/genes_v05/'

path.data = '../01_data/'

```

## Make softlinks
find /Volumes/Samsung_T5/vienn/annotation/  -name "*.gff" -type f -exec ln -s {} ./ \;

## Get all gff files
```{r}

gff.files <- list.files(path = path.raw, pattern = "\\.gff$", full.names = F)

accessions <- gsub('.gff', '', gff.files)
accessions = setdiff(accessions, '0')
accessions = setdiff(accessions, '22001')

length(accessions)

```


# Gff to gff
```{r}

path.cons = '/Volumes/Samsung_T5/vienn/00_paper/01_data/02_alignment/pannagram_v05/'


gff.all = c()
for(acc in accessions){
  
  gff1 = read.table(paste(path.raw, acc, '.gff', sep = ''), stringsAsFactors = F)
  
  gff2 = gff2gff(path.cons = path.cons, acc1 = acc, acc2 = 'PanGen',
        gff1 = gff1, aln.type = 'v_', ref.acc='0', exact.match=F)
  
  gff.all = rbind(gff.all, gff2)

}

head(gff.all)

gff.all$V1 = gsub('pangen_Chr', 'PanGen_Chr_', gff.all$V1)
gff.all$V1 = gsub('PanGen_Chr', 'PanGen_Chr_', gff.all$V1)

gff.all$V1 = gsub('PanGen_Chr__', 'PanGen_Chr_', gff.all$V1)

# saveRDS(gff.all, paste(path.pan.raw, 'gff_raw_pan_v05.rds', sep = ''), compress = F)

write.table(gff.all[gff.all$V3 == 'exon',1:9], paste(path.pan.raw, 'gff_raw_pan_v05_exon.gff', sep = ''),
            quote = F, col.names = F, row.names = F, sep = '\t')


head(gff1)


```



# Count gene split and gene merge
```{r}

names.split = c()
names.merge = c()
for(acc in accessions){
  gff.raw = read.table(paste(path.raw, acc, '.gff', sep = ''), stringsAsFactors = F)
  gff.cleaned = read.table(paste(path.pan.genes,
                                 'genes_v05_', acc,'.gff', sep = ''), stringsAsFactors = F)
  gff.raw = gff.raw[gff.raw$V3 == 'gene',]
  gff.cleaned = gff.cleaned[gff.cleaned$V3 == 'gene',]
  
  # Gene names
  gff.cleaned$names = sapply(gff.cleaned$V9, function(s) strsplit(s, ';')[[1]][1])
  
  
  # Remove non-chromosomes
  gff.raw = gff.raw[grepl('Chr', gff.raw$V1),]
  gff.cleaned = gff.cleaned[grepl('Chr', gff.cleaned$V1),]
  
  # Estimate the chromosome number
  gff.raw$chr = as.numeric(sapply(gff.raw$V1, function(s) strsplit(s, 'Chr')[[1]][2]))
  gff.cleaned$chr = as.numeric(sapply(gff.cleaned$V1, function(s) strsplit(s, 'Chr')[[1]][2]))
  
  n.merge = 0
  n.split = 0
  for(i.chr in 1:5){
    pokaz('Chr', i.chr)
      for(s.strand in c('+', '-')){
        gff.raw.s = gff.raw[(gff.raw$V7 == s.strand) & (gff.raw$chr == i.chr),]
        gff.cleaned.s = gff.cleaned[(gff.cleaned$V7 == s.strand) & (gff.cleaned$chr == i.chr),]
        gff.cleaned.s = gff.cleaned.s[order(gff.cleaned.s$V4),]
        
        
        while(T){
          tmp = gff.cleaned.s$V4[-1] - gff.cleaned.s$V5[-nrow(gff.cleaned.s)]
          tmp = which(tmp < 0)
          if(length(tmp) == 0)
            break
          gff.cleaned.s = gff.cleaned.s[-tmp,]
        }

        pos.raw = rep(0, 35000000)
        pos.raw[gff.raw.s$V4] = 1:nrow(gff.raw.s)
        pos.raw[gff.raw.s$V5] = -(1:nrow(gff.raw.s))
        pos.raw = cumsum(pos.raw)
        pos.raw[gff.raw.s$V5] = (1:nrow(gff.raw.s))
      
        if(sum(pos.raw != 0) != sum(gff.raw.s$V5 - gff.raw.s$V4 + 1)){
          stop('Wrong coorsinates in Raw')
        } 
        
        pos.cleaned = rep(0, 35000000)
        pos.cleaned[gff.cleaned.s$V4] = 1:nrow(gff.cleaned.s)
        pos.cleaned[gff.cleaned.s$V5] = -(1:nrow(gff.cleaned.s))
        pos.cleaned = cumsum(pos.cleaned)
        pos.cleaned[gff.cleaned.s$V5] = (1:nrow(gff.cleaned.s))
        
        if(sum(pos.cleaned != 0) != sum(gff.cleaned.s$V5 - gff.cleaned.s$V4 + 1)){
          stop('Wrong coorsinates in Cleaned')
        } 
        
        idx.both = which(pos.cleaned * pos.raw != 0)
        
        # Gene merge
        n.raw.in.clean = tapply(pos.raw[idx.both], pos.cleaned[idx.both], function(x) length(unique(x)))
        n.merge = n.merge + sum(n.raw.in.clean > 1)
        
        names.merge = c(names.merge, 
                        gff.cleaned.s$names[as.numeric(names(n.raw.in.clean)[n.raw.in.clean > 1])])
        
        # Gene split
        n.cleaned.in.raw = tapply(pos.cleaned[idx.both], pos.raw[idx.both], function(x) list(unique(x)))
        n.split = n.split + sum(lapply(n.cleaned.in.raw, length) > 1)
        
        
        names.split = c(names.split, 
                        gff.cleaned.s$names[unlist(n.cleaned.in.raw[lapply(n.cleaned.in.raw, length) > 1])])
        
    }
  }

  
}

saveRDS(unique(names.split), paste(path.data, 'names_split.rds', sep = ''))
saveRDS(unique(names.merge), paste(path.data, 'names_merge.rds', sep = ''))

length(unique(names.split))
length(unique(names.merge))


```

