---
title: "R Notebook"
output: null_document
---



# Setup
## My Libs
```{r}

path.base = '../../../'

path.pannagram = paste(path.base, '../pannagram/', sep = '')

source(paste(path.pannagram, 'sim/sim_func.R', sep = ''))
source(paste(path.pannagram, 'sv/graph_refinement.R', sep = ''))
source(paste(path.pannagram, 'utils/utils.R', sep = ''))
source(paste(path.pannagram, 'visualisation/dotplot.R', sep = ''))
source(paste(path.pannagram, 'visualisation/msaplot.R', sep = ''))
source(paste(path.pannagram, 'visualisation/orfplot.R', sep = ''))
source(paste(path.pannagram, 'visualisation/visualisation.R', sep = ''))
```


## Libs
```{r}
invisible(library(ggplot2))
invisible(library(rhdf5))
```


# MSA
## Read 
```{r}
path.gaps = '/Volumes/Samsung_T5/vienn/alignment/new/mafft_out2/mafft_out/'
file = 'Gap_5_5_79155_22618867_22624777_flank_30_aligned.fasta'  # 1
file = 'Gap_2_2_08781_2372098_2381871_flank_30_aligned.fasta'


file = 'Gap_3_3_34942_10687357_10688859_flank_30_aligned.fasta'


file.path = paste(path.gaps, file, sep = '')


# Read the alignemnt
aln = readFastaMy(file.path)
mx = toupper(aln2mx(aln))
msaplot(mx)
```

## Refine
```{r}
n.flank = 30
res = refineMafft(mx, n.flank = n.flank)
```

## Find clusters

```{r}
mx.cut = res$mx
mx.common = mx.cut[,colSums(mx.cut != '-') >= 2]

msaplot(mx.common)

mx.common.dist = mx2dist(mx.common, ratio = T)

hc = hclust(d = as.dist(mx.common.dist))
plot(hc)

threshold = 0.3
clusters = cutree(hc, h = threshold)
clusters = sort(clusters)

table(clusters)

msaplot(mx[names(clusters),])

```

# Align every cluster separately
```{r}


 # !!!!!!!построить дерево с помощью mash

Sys.setenv(PATH = paste("/Users/annaigolkina/miniconda3/envs/bioinf/bin", Sys.getenv("PATH"), sep=":"))


seqs = mx2seq(mx)
path.mattf.tmp = '/Volumes/Samsung_T5/vienn/alignment/new/mafft_tmp/tmp/'
pref = 'tmp4.fasta'

s.cons.all = c()
mx.all = list()
for(i.cl in 1:max(clusters)){
  pokaz('Cluster', i.cl)
  s = seqs[clusters == i.cl]
  
  if(sum(clusters == i.cl) == 1){
    s.cons.all[i.cl] = s
    mx.all[[i.cl]] = aln2mx(s[1])
    next
  }
  
  file.tmp = paste(path.mattf.tmp,pref, '_cl_', i.cl,'.fasta', sep = '')
  file.out = paste(path.mattf.tmp,pref, '_cl_', i.cl,'_out.fasta', sep = '')
  writeFastaMy(s, file.tmp)
  
  command <- paste('mafft --op 10 --quiet --ep 0.1  ',file.tmp,' > ',file.out, sep = '')
  system(command)
  
  
  aln.cl = readFastaMy(file.out)
  mx.cl = aln2mx(aln.cl)
  msaplot(mx.cl)
  
  msadiff(mx.cl)
  
  # CLEAN UP THE ALIGNMENT!!!!!
  
  mx.all[[i.cl]] = mx.cl
  
  s.cons.all[i.cl] = nt2seq(mx2cons(mx.cl))
  
  command <- paste('rm ', file.tmp, sep = '')
  system(command)
  command <- paste('rm ', file.out, sep = '')
  system(command)
}

file.tmp = paste(path.mattf.tmp,pref, '_cl_cons.fasta', sep = '')
file.out = paste(path.mattf.tmp,pref, '_cl_cons_out.fasta', sep = '')
names(s.cons.all)  = paste('cluster_', 1:length(s.cons.all), sep = '')
writeFastaMy(s.cons.all, file.tmp)

command <- paste('mafft --op 10 --quiet --ep 0.1  ',file.tmp,' > ',file.out, sep = '')
system(command)


aln.cl = readFastaMy(file.out)
mx.cl = aln2mx(aln.cl)
msaplot(mx.cl)


dotplot.s(s.cons.all[1], s.cons.all[2], 15, 12)



```

# Find blocks with poor alignment
```{r}


# Diversity by each position
mx.len = ncol(mx.cl)
mx.cl = toupper(mx.cl)
s.nts = c('A', 'C', 'G', 'T')
pos.profile = matrix(0, nrow = 4, ncol = mx.len, dimnames = list(c(s.nts, NULL)))
for(s.nt in s.nts){
  pos.profile[s.nt,] = colSums(mx.cl == s.nt)
}
pos.variation = (colSums(pos.profile == 0) != 3) * 1


plot(pos.variation)



sim.cutoff = 0.3

wnd.len = max(round(mx.len * 0.01), 15)
n.mismatch = round(wnd.len * sim.cutoff)


d = pos.variation

d.sum <- d
for (i in 1:(wnd.len - 1)) {
  # Сдвигаем вектор
  d.shift <- lag(d, n = i, default = 0)
  # Суммируем с исходным вектором сумм
  d.sum <- d.sum + d.shift
}
d.sum = d.sum[-(1:(wnd.len-1))]


plot(d.sum)

```





```{r}

msaplot(mx)
msaplot(res$mx)

```

mafft --op 3  --ep 0.1 --treein tree.mafft  tmp4.fasta > mm4.fasta



```{r}
mx2 = res$mx
mx2 = mx2[,colSums(mx2 != '-') >= 2]

msaplot(mx2)

mx.dist = mx2dist(mx2, ratio = T)

new.names = paste('s', 1:nrow(mx), sep = '')

colnames(mx.dist) = new.names
rownames(mx.dist) = new.names

seqs = mx2seq(res$mx)
names(seqs) = new.names

writeFastaMy(seqs, '/Volumes/Samsung_T5/vienn/alignment/new/mafft_tmp/seqs4.fasta')


library(ape)

hc = hclust(d = as.dist(mx.dist))
plot(hc)

threshold = 0.3
clusters = cutree(hc, h = threshold)

clusters = sort(clusters)
table(clusters)

msaplot(mx[names(clusters),])

# two seqeunces for mafft
seqs = mx2seq(res$mx)
seqs2 = seqs[names(clusters)[c(7: 23)]]
writeFastaMy(seqs2, '/Volumes/Samsung_T5/vienn/alignment/new/mafft_tmp/seqs4.fasta')



phylo_tree <- as.phylo(hc)
phylo_tree$edge.length <- NULL
write.tree(phylo_tree, file = "/Volumes/Samsung_T5/vienn/alignment/new/mafft_tmp/tree.nwk")

```


## Clustering left and right
```{r}


num.leaves <- length(hc$order)
num.nodes <- nrow(hc$merge)

# Инициализация матрицы
mx.leaf <- matrix(0, nrow = num.nodes, ncol = num.leaves)

for(i.node in 1:num.nodes){
  clade.p = hc$merge[i.node, 1]  # assign as +1
  clade.m = hc$merge[i.node, 2]  # assign as -1
  if(clade.p < 0){  # read node
    mx.leaf[i.node, abs(clade.p)] = 1
  } else {
    mx.leaf[i.node, abs(mx.leaf[clade.p,]) > 0] = 1
  }
  
  if(clade.m < 0){  # read node
    mx.leaf[i.node, abs(clade.m)] = -1
  } else {
    mx.leaf[i.node, abs(mx.leaf[clade.m,]) > 0] = -1
  }
  
}

if(rowSums(abs(mx.leaf))[num.nodes] != num.leaves) stop('Wrong matrix')

hc$height


```


```{r}
aln = readFastaMy('/Volumes/Samsung_T5/vienn/alignment/new/mafft_tmp/aa4.fasta')
mx0 = aln2mx(aln)

s.cons = nt2seq(mx2cons(mx0))


dotplot.s(s.cons, seqs[10], 15, 12)

msaplot(mx0)

msadiff(mx0)


# dotplot.s(seqs[names(clusters)[7]], seqs[names(clusters)[6]], 15, 12)


```


```{r}

sim.cutoff = 0.3

wnd.len = max(round(ncol(mx0) * 0.01), 15)
n.mismatch = round(wnd.len * sim.cutoff)


d = (mx0[1,] != mx0[2,]) * 1

d.sum <- d
for (i in 1:(wnd.len - 1)) {
  # Сдвигаем вектор
  d.shift <- lag(d, n = i, default = 0)
  # Суммируем с исходным вектором сумм
  d.sum <- d.sum + d.shift
}
d.sum = d.sum[-(1:(wnd.len-1))]

which(d.sum <= n.mismatch)


plot(d.sum <= n.mismatch)


```



```{r}

s1 = mx2seq(mx)
s2 = mx2seq(res$mx)

for(i in 1:length(s1)){
  pokaz(s1[1] == s2[1])
}

dotplot.s(s1[1], s2[1], 15, 12)


```



```{r}
p0
p2
p4
```

```{r}
path.gaps = '/Volumes/Samsung_T5/vienn/alignment/new/mafft_out2/mafft_out/'
# file = 'Gap_1_1_007260_2409857_2410708_flank_30_aligned.fasta'
file = 'Gap_5_5_79155_22618867_22624777_flank_30_aligned.fasta'  # 1
file = 'Gap_1_1_068932_21522433_21533516_flank_30_aligned.fasta' # 2
file = 'Gap_2_2_13563_5079699_5097933_flank_30_aligned.fasta' # ok
file = 'Gap_4_4_06465_2229137_2241635_flank_30_aligned.fasta'  # 4

file.path = paste(path.gaps, file, sep = '')


# file.path = '/Volumes/Samsung_T5/vienn/alignment/new/mafft_tmp/aln2.fasta'


# Remove flanking regions
n.flank = 30

# Read the alignemnt
aln = readFastaMy(file.path)
mx = toupper(aln2mx(aln))

mx.pos = matrix(0, nrow = nrow(mx), ncol = ncol(mx), 
                     dimnames = list(rownames(mx), NULL))
for(irow in 1:nrow(mx.pos)){
  pos.row = as.numeric(strsplit(rownames(mx)[irow], '\\|')[[1]][3:4])
  pos.row = pos.row[1]:pos.row[2] 
  pos.row = 1:length(pos.row)
  
  pos.filled = which(mx[irow,] != '-')
  pos.filled = pos.filled[-(1:n.flank)]
  pos.filled <- pos.filled[-((length(pos.filled) - n.flank + 1):length(pos.filled))]
  
  mx.pos[irow, pos.filled] = pos.row
}

# Remove flanking
mx[mx.pos == 0] = '-'
idx.col.remain = (colSums(mx != '-') != 0)
mx = mx[,idx.col.remain]
mx.pos = mx.pos[,idx.col.remain]

# Save variables
mx.init = mx
mx.pos.init = mx.pos

# Plot
p = msaplot(mx)
p
pdf(paste('/Volumes/Samsung_T5/vienn/alignment/new/mafft_tmp/', 'tmp4.pdf', sep = ''), width = 10, height = 4)
print(p)     # Plot 1 --> in the first page of PDF
dev.off()
```

## Fplit
```{r}
sim.cutoff = 0.3
s.nt.fake = '!'

# Final alignment
mx.ok = c()
mx.pos.ok = c()

# To arrange a cycle
mx.rest = mx.init
mx.pos.rest = mx.pos.init

for(n.round in 1:5){
  pokaz('Disentangle Round', n.round)
  
  mx = mx.rest           # The matrix you work with and change values
  mx.pos = mx.pos.rest   # The matrix you work with and change values
  
  # Diversity by each position
  aln.len = ncol(mx)
  mx = toupper(mx)
  s.nts = c('A', 'C', 'G', 'T')
  pos.profile = matrix(0, nrow = 4, ncol = aln.len, dimnames = list(c(s.nts, NULL)))
  for(s.nt in s.nts){
    pos.profile[s.nt,] = colSums(mx == s.nt)
  }
  pos.variation = (colSums(pos.profile == 0) != 3) * 1
  
  # Define blocks, were the alignment non well
  blocks.all = c()
  for(i.seq in 1:nrow(mx)){
    s = mx[i.seq,]
    blocks = findOnes((s != '-') *1)
    blocks$len = blocks$end - blocks$beg + 1
    if(nrow(blocks) == 0) next
    
    # Estimate diversity within each block
    blocks$pi = 0
    blocks$acc = rownames(mx)[i.seq]
    for(irow in 1:nrow(blocks)){
      idx.block = blocks$beg[irow]:blocks$end[irow]
      blocks$pi[irow] = sum(pos.variation[idx.block]) / blocks$len[irow]
    }
    
    blocks.all = rbind(blocks.all, blocks)
  }
  if(nrow(blocks) == 0){
    pokaz('Exit 1')
    mx.ok = cbind(mx.ok, mx)
    mx.pos.ok = cbind(mx.pos.ok, mx.pos)
    break
  } 
  
  blocks = blocks.all
  rm(blocks.all)
  blocks = blocks[blocks$pi>sim.cutoff,,drop = F]
  blocks = blocks[order(-blocks$pi),,drop = F]
  
  if(nrow(blocks) == 0){
    pokaz('Exit 2')
    mx.ok = cbind(mx.ok, mx)
    mx.pos.ok = cbind(mx.pos.ok, mx.pos)
    break
  }
  
  blocks$remove = 0
  for(irow in 1:nrow(blocks)){
    idx.block = blocks$beg[irow]:blocks$end[irow]
    blocks$pi[irow] = sum(pos.variation[idx.block]) / blocks$len[irow]
    if(blocks$pi[irow] <= sim.cutoff){
      # stop()
      # pokaz(irow)
      next
    } 
    
    seq.tmp = mx[blocks$acc[irow], idx.block]  
    for(s.nt in s.nts){
      pos.profile[s.nt,idx.block] = pos.profile[s.nt,idx.block] - 1 * (seq.tmp == s.nt)
    }
    
    blocks$remove[irow] = 1
    pos.variation[idx.block] = (colSums(pos.profile[,idx.block,drop = F] == 0) != 3) * 1
  }
  
  
  # # Optional
  # for(irow in 1:nrow(blocks)){
  #   idx.block = blocks$beg[irow]:blocks$end[irow]
  #   loneliness = max(pos.profile[,idx.block,drop=F])
  #   if(loneliness == 1){
  #     # pokaz(irow)
  #     # stop()
  #     blocks$remove[irow] = 1
  #   } 
  # }
  
  
  
  # Remove blocks with very high diversity
  for(irow in 1:nrow(blocks)){
    if(blocks$remove[irow] == 0) next
    idx.block =  blocks$beg[irow]:blocks$end[irow]
    mx[blocks$acc[irow],idx.block] = s.nt.fake
  }
  
  
  # Well alignmed positions
  mask.ok = (mx != s.nt.fake)
  mx[!mask.ok] = '-'
  mx.pos[!mask.ok] = 0
  
  mx.ok = cbind(mx.ok, mx)
  mx.pos.ok = cbind(mx.pos.ok, mx.pos)
  
  # The rest for the next round
  
  mx.rest[mask.ok] = '-'
  mx.pos.rest[mask.ok] = 0
  
  cols.rest = colSums(mx.rest != '-') > 0
  mx.rest = mx.rest[, cols.rest, drop = F]
  mx.pos.rest = mx.pos.rest[, cols.rest, drop = F]
  
  if(ncol(mx.rest) == 0){
    pokaz('Exit 3')
    break
  }
  
}
if(sum(rowSums(mx.pos.ok != 0) != rowSums(mx.pos.init != 0)) != 0) stop('Not enough rounds')
if(sum(dim(mx.ok) != dim(mx.pos.ok)) != 0) stop('Sizes of matrices are different')

p2 = msaplot(mx.ok)
p2

# p3 = msaplot(mx.rest)
# p3




```


## Ordering
```{r}
mx = mx.ok
mx.pos = mx.pos.ok

# Remove total gaps
idx.remain = colSums(mx != '-') > 0
mx = mx[,idx.remain]
mx.pos = mx.pos[,idx.remain]
mx.len = ncol(mx)


# blocks of connetedness
b.beg = 1
b.end = c()
mx.beg = c()
mx.end = c()

val.beg = mx.pos[,b.beg]
val.end = mx.pos[,b.beg]
for(icol in 2:mx.len){
  val.nex = mx.pos[,icol]
  
  d = val.nex - val.end
  d = d[(val.end != 0) & (val.nex != 0)]
  if(length(d) == 0) d = 0
  
  if(sum(d != 1) == 0){
    val.beg[val.beg == 0] = val.nex[val.beg == 0]
    val.end[val.nex != 0] = val.nex[val.nex != 0]
  } else {
    # stop()
    b.end = c(b.end, icol-1)
    b.beg = c(b.beg, icol)
    
    mx.beg = cbind(mx.beg, val.beg)
    mx.end = cbind(mx.end, val.end)
    
    if(sum(val.beg == 0) != sum(val.beg * val.end == 0)) stop()
    
    val.beg = val.nex
    val.end = val.nex
  }
}
b.end = c(b.end, mx.len)  # last ending
mx.beg = cbind(mx.beg, val.beg)
mx.end = cbind(mx.end, val.end)

# Common data.frame
blocks = data.frame(beg = b.beg, end = b.end)
blocks$len = blocks$end - blocks$beg + 1
```


```{r}
# Sorting
n = nrow(blocks)
index <- 1:n 
i <- 2
while (i <= n) {

  beg.i <- mx.beg[,index[i]]
  end.i <- mx.end[,index[i]]
  
  key.i <- index[i]
  k <- 0  # position which fit the
  for (j in (i - 1):1) {
    beg.j = mx.beg[,index[j]]
    end.j = mx.end[,index[j]]
    
    idx.common = (beg.j * beg.i) != 0
    
    if(sum(idx.common) == 0){
      next
    } else if(max(abs(beg.j[idx.common] - end.i[idx.common])) == 1) {
      k = j
      break
    } else if(max(abs(beg.i[idx.common] - end.j[idx.common])) == 1) {
      k = j+1
      break
    }else if(sum(x[idx.common] > y[idx.common])) {
      break  
    }
  }
  # stop()
  if(k == 0){
    # stop('skip')
    tmp = index[i]
    index[i:(n-1)] = index[(i+1):n]
    index[n] = tmp
    next # not to increase the i index
  } else if (k != i) {
    # stop('middle fit')
    tmp = index[i]
    index[(k+1):i] = index[k:(i-1)]
    index[k] = tmp
    
  } else {
    # stop('stay')
  }
  i <- i + 1  
  
}

pos = c()
for(i in index){
  pos = c(pos, blocks$beg[i]:blocks$end[i])
}


p4 = msaplot(mx[,pos])
p4

pdf(paste('/Volumes/Samsung_T5/vienn/alignment/new/mafft_tmp/', 'tmp4_blocks.pdf', sep = ''), width = 10, height = 4)
print(p4)     # Plot 1 --> in the first page of PDF
dev.off()


# Check that every sequence is sorted
for(irow in 1:nrow(mx.pos)){
  p = mx.pos[irow,pos]
  p = p[p != 0]
  if(is.unsorted(p)) print(irow)
  
  # which( mx.pos[irow,] %in% which(diff(p) !=1))
}


aln.new = mx2aln(mx[,pos])

writeFastaMy(aln.new, '/Volumes/Samsung_T5/vienn/alignment/new/mafft_tmp/im.fasta')


tmp = readFastaMy('/Volumes/Samsung_T5/vienn/alignment/new/mafft_tmp/im2.fasta')
mx2 = aln2mx(tmp)
p1 = msaplot(mx2)
p1


mx0 = mx.init[names(tmp),]
p0 = msaplot(mx0)
p0


mx1 = mx[names(tmp),pos]
p1 = msaplot(mx1)
p1

```



```{r}
pdf(paste('/Volumes/Samsung_T5/vienn/alignment/new/mafft_tmp/', 'tmp4.pdf', sep = ''), width = 10, height = 4)
print(p2)     # Plot 1 --> in the first page of PDF
dev.off()


msadiff(mx)

```

## tree
```{r}
# install.packages("ape")
library("ape")


tree <- read.tree("/Volumes/Samsung_T5/vienn/alignment/new/mafft_tmp/tmp4.fasta.tree")  # Замените path_to_your_file.nwk на путь к вашему файлу

# Визуализация дерева
plot(tree)

```







```{r}
seqs = aln
for(i in 1:length(seqs)){
  s = seq2nt(seqs[i])
  seqs[i] = nt2seq(s[s != '-'])
}

# dotplot.s(seqs[1], seqs[2], 15, 12)
# 
# names(seqs)[2]
# names(seqs)[6]
# 

writeFastaMy(seqs, '/Volumes/Samsung_T5/vienn/alignment/new/mafft_tmp/tmp1.fasta')


```


```{r}

acc = '6124'
i.chr = 1
s1 = seq2nt(readFastaMy(paste('/Volumes/Samsung_T5/vienn/pb_chromosomes/',acc,'_chr',i.chr,'.fasta', sep = '')))
s2 = seq2nt(readFastaMy(paste('/Volumes/Samsung_T5/vienn/pb_chromosomes/0_chr',i.chr,'.fasta', sep = '')))


i.inv = 2
beg1 = 20764752
end1 = 20766539
beg2 = 21522434
end2 = 21533515

p = dotplot(s1[(beg1-1000):(end1+1000)], s2[(beg2-1000):(end2+1000)], 15, 12)
p


```




