---
title: "SV comparissons"
output: null_document
---
# Utils
```{r}
source('../../../../pannagram/utils/utils.R')
library(rhdf5)

```


## Additional variables
```{r}

gr.accs.e <- "accs/"
gr.accs.b <- "/accs"

```

# Functions
```{r}
visualiseDiff <- function(pos1, pos2, s.pref = '', diff.mode = F){
  
  idx.mess.panna = ((pos.beg <= pos1) & (pos1 <= pos.end)) | 
    ((pos.beg <= pos2) & (pos2 <= pos.end)) |
    ((pos1 <= pos.beg) & (pos.end <= pos2))
  
  mess.panna = data.frame(beg = pos.beg[idx.mess.panna], end = pos.end[idx.mess.panna])
  
  idx.mess.graph =  ((sv.graph$V2 <= pos1) & (pos1 <= sv.graph$V3)) | 
    ((sv.graph$V2 <= pos2) & (pos2 <= sv.graph$V3)) |
    ((pos1 <= sv.graph$V2) & (sv.graph$V3 <= pos2))
    
    
    (pos1 <= sv.graph$V2) & (pos2 >= sv.graph$V3)
  mess.graph = data.frame(beg = sv.graph$V2[idx.mess.graph], end = sv.graph$V3[idx.mess.graph])
  
  
  p1 = orfplot(mess.panna) + geom_vline(xintercept = c(pos1, pos2))
  p2 = orfplot(mess.graph) + geom_vline(xintercept = c(pos1, pos2))
  
  
  # Create the MSA
  path.cons = '/Volumes/Samsung_T5/vienn/msa/'
  ref.acc = '0'
  
  i.chr = 1
  
  # Get positions in the pangenome coordinates
  file.msa = paste(path.cons, 'v_', i.chr, '_', i.chr, '_ref_',ref.acc,'.h5', sep = '')
  v.acc = h5read(file.msa, paste(gr.accs.e, acc, sep = ''))
  pos1.acc = which(v.acc == pos1)
  pos2.acc = which(v.acc == pos2)
  
  if((length(pos1.acc) == 0) || (length(pos2.acc) == 0)){
    pokazAttention('Wrong positions')
    return()
  }
  
  
  # Get Alignment
  file.seq.msa = paste(path.cons, 'seq_', i.chr, '_', i.chr, '_ref_',ref.acc,'.h5', sep = '')
  
  h5ls(file.seq.msa)
  
  # Get accession names
  groups = h5ls(file.seq.msa)
  accessions = groups$name[groups$group == gr.accs.b]
  
  # Initialize vector and load MSA data for each accession
  seq.mx = matrix('-', nrow = length(accessions), ncol = pos2.acc - pos1.acc + 1)
  for(i.acc in 1:length(accessions)){
    pokaz(accessions[i.acc])
    s.acc = h5read(file.seq.msa, paste(gr.accs.e, accessions[i.acc], sep = ''))
    seq.mx[i.acc,] = s.acc[pos1.acc:pos2.acc]
  }
  rownames(seq.mx) = accessions
  
  
  
  if(diff.mode){
    p3 = msadiff(seq.mx)
    s.diff = '_diff'
  } else {
    p3 = msaplot(seq.mx)
    s.diff = ''
  }
  
  
  
  
  
  
pdf(paste(path.figures.comp, s.pref, '_', acc, '_chr_', i.chr, '_', 
          formatC(pos1, format = "d"), '_', 
          formatC(pos2, format = "d"),s.diff, '.pdf', sep = ''), width = 8, height = 8)

  grid.arrange(
    arrangeGrob(p1, p2, ncol = 2), # Устанавливаем верхние графики в один ряд
    p3,                            # Нижний график
    heights = c(1, 2)              # Относительные высоты для верхних и нижнего графиков
  )
  dev.off()
}
```


# Setup
```{r, message=FALSE}

if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager", quiet = TRUE)

if (!requireNamespace("Biostrings", quietly = TRUE)) {
    BiocManager::install("Biostrings", quiet = TRUE)
}

if (!require("ggnet", character.only = TRUE, quietly = TRUE)) {
    load.library("devtools")
    devtools::install_github("briatte/ggnet")
}

packages <- c("ggpattern", 
              "viridis", 
              "colorRamps", 
              "gridExtra", 
              "ggplot2", 
              "igraph", 
              "network", 
              "khroma", 
              "dplyr", 
              "ggpubr", 
              "Biostrings",
              "ggnet",
              "patchwork")

invisible(sapply(packages, load.library))



# source('similarity.R')
# source('graph_refinement.R')
source('../../../../pannagram/visualisation/msaplot.R')
source('../../../../pannagram/visualisation/dotplot.R')
source('../../../../pannagram/visualisation/orfplot.R')


sunset <- colour("sunset")
discrete_rainbow <- colour("discrete rainbow")

path.base = '../../../'
path.work = paste(path.base, '02_analysis/04_sv/01_data/', sep = '')
path.tair = paste(path.base, '01_data_common/01_tair10/', sep = '')
path.figures = paste(path.base, '02_analysis/04_sv/03_figures/', sep = '')
path.sv.panna = paste(path.base, '01_data_common/02_annot_denovo/02_pannagram/svs/', sep = '')
path.sv.graph = '/Users/annaigolkina/Downloads/sebastian_svs/'


min.len = 15
max.len = 20000

i.chr = 1
sv.anna.beg = read.table(paste(path.sv.panna, 'svs_all_beg_pos_v03.txt', sep = ''), check.names = F)
sv.anna.end = read.table(paste(path.sv.panna, 'svs_all_end_pos_v03.txt', sep = ''), check.names = F)


```


# Get positions
## Pannagram
```{r}


regex <- paste0("^SVgr_", i.chr)
idx.chr <- grep(regex, rownames(sv.anna.beg))
sv.anna.beg = sv.anna.beg[idx.chr,]
sv.anna.end = sv.anna.end[idx.chr,]

acc = '1741'

chr.len = 31 * 10^6

pos.graph = rep(0, chr.len)


pos.beg = sv.anna.beg[, acc]
pos.end = sv.anna.end[, acc]
pos.len = abs(pos.end - pos.beg) - 1

idx.remain = ((pos.beg != 0) & (pos.end != 0) & (pos.len != 0))
idx.remain = ((pos.beg != 0) & (pos.end != 0) & (pos.len >= min.len))
idx.remain = ((pos.beg != 0) & (pos.end != 0) & (pos.len >= min.len) & (pos.len <= max.len))
idx.tmp = pos.beg > pos.end
tmp = pos.beg[idx.tmp]
pos.beg[idx.tmp] = pos.end[idx.tmp]
pos.end[idx.tmp] = tmp

pos.beg = pos.beg[idx.remain] + 1
pos.end = pos.end[idx.remain] - 1
pos.len = pos.len[idx.remain]
idx.more1 = (pos.len != 1)

pos.panna = rep(0, chr.len)
for(irow in 1:length(pos.beg)){
  pos.panna[pos.beg[irow]:pos.end[irow]] = 1
}





```

## Graph
```{r}

sv.graph = read.table(paste(path.sv.graph, "chr", i.chr,".",acc, "_Chr",i.chr,".bed", sep = ''), stringsAsFactors = F,
                      comment.char = '$')
sv.graph$len = sv.graph$V3 - sv.graph$V2
sv.graph = sv.graph[(sv.graph$len >= min.len) & (sv.graph$len <= max.len),]

pos.graph = rep(0, chr.len)
for(irow in 1:nrow(sv.graph)){
  pos.graph[(sv.graph$V2[irow]+1):(sv.graph$V3[irow])] = 1
}

sum(pos.graph)


# plot(pos.graph)
```


# Differences
## Contigency table
```{r}
table(pos.graph, pos.panna)
```

# Along the chromosome
```{r}
pos.diff = (pos.graph>0) & (pos.panna == 0)  | 
  (pos.graph == 0) & (pos.panna != 0)


wnd.diff = wndMean(pos.diff)

wnd.panna = wndMean((pos.panna != 0) * 1)


plot(wnd.diff)
plot(wnd.panna, wnd.diff)

plot(wnd.panna)

sum(wnd.panna * 10000)
  

df = data.frame(x = 1:length(wnd.diff), y = wnd.diff)
ggplot(df, aes(x = x, y = y)) + 
    geom_bin2d(bins = c(50, 20)) +
    scale_fill_gradient(low = "#F8FAE5", high = "#D04848") +
    geom_point(color = "black", size = 0.5, alpha = 0.3) +
    theme_minimal() + xlim(c(0,3000)) +
  geom_smooth(color = "#7D0A0A", se = FALSE) +
  xlab('# of window') + ylab('ratio of diff between Pannagram and Graph')




```



# Visualise
## Diff by percentage
```{r}
diff.cutoff = 0.4
idx.mess = which(wnd.diff > diff.cutoff)
pokaz('Number of windows with', diff.cutoff, 'is', length(idx.mess))
for(k in 1:length(idx.mess)){
  pokaz('Window number', k)
  # k = 1
  pos1 = wnd.size * (idx.mess[k]-1) + 1
  pos2 = wnd.size * (idx.mess[k])
  pokaz('Positions', pos1, pos2)
  
  sum(pos.diff[pos1:pos2])
  
  visualiseDiff(pos1, pos2, s.pref = 'panna', diff.mode = T)  
}

```

## Sebastian cases
```{r}
# ----------

path.figures.comp = '../04_examples/comp_panna_graph/figures/'
if (!dir.exists(path.figures.comp)) {
  dir.create(path.figures.comp)
}


file.positions = '../04_examples/comp_panna_graph/graph_examples_1741.txt'
positions = read.table(file.positions, stringsAsFactors = F)
for(i.pos in 1:nrow(positions)){
  pos1 = positions$V2[i.pos]
  pos2 = positions$V3[i.pos]  

  visualiseDiff(pos1, pos2)


}

```




